# AQL形式（Academic Quiz League）仕様書

## 概要

AQL形式（Academic Quiz League）は、Score Watcherで利用できるチーム戦形式の競技クイズルールです。Academic Quiz League（アカデミック・クイズ・リーグ）で実際に使用されている形式を再現しており、10人のプレイヤーが2つのチームに分かれて競技します。

## 基本仕様

### ゲームルール名

- **システム名**: `aql`
- **表示名**: AQL
- **短縮説明**: クイズ大会「AQL」で使われている形式です。
- **詳細説明**: 10人のプレイヤーが2つのチームに分かれ、プレイヤーのスコアの積が200を超えたチームが優勝です。

### ゲーム設定項目

| 項目         | 設定値の例 | 説明                                           |
| ------------ | ---------- | ---------------------------------------------- |
| `options`    | object     | チーム名設定                                   |
| `left_team`  | "Team A"   | 左チーム（プレイヤー1-5番）の表示名           |
| `right_team` | "Team B"   | 右チーム（プレイヤー6-10番）の表示名          |

**重要な制約**：
- プレイヤー数は**必ず10人固定**です
- プレイヤー1-5番が左チーム、プレイヤー6-10番が右チームに自動的に割り当てられます

## スコア計算ロジック

### 初期状態

各プレイヤーの初期状態：

- **スコア**: 1（他の形式と異なり、0ではなく1から開始）
- **正解数**: 0
- **誤答数**: 0
- **状態**: `playing`（プレイ中）
- **失格状態**: `false`（失格していない）

### 正解時の処理

1. プレイヤーのスコアを1増加（スコア = 正解数 + 1）
2. 正解数を1増加
3. 最後の正解問題番号を記録
4. チーム勝利条件を確認

### 誤答時の処理

AQL形式では特殊な誤答処理システムが採用されています：

#### 1回目の誤答
- 誤答数を1増加
- スコアを1にリセット
- 状態を「誤答リーチ」に変更
- 最後の誤答問題番号を記録

#### 2回目の誤答（失格確定）
- 誤答数を2に増加
- スコアを1にリセット
- 状態を「失格」（`lose`）に変更
- 失格フラグ（`is_incapacity`）を`true`に設定

### 復活システム

AQL形式の最大の特徴である**自動復活システム**：

**復活条件**：
- 失格状態のプレイヤーがいる場合
- **相手チーム**のプレイヤーが誤答した時
- 即座に自動復活

**復活時の処理**：
- スコアを1にリセット
- 状態を「プレイ中」（`playing`）に戻す
- 失格フラグを`false`に設定
- 誤答リーチ状態も解除

**チーム判定**：
- プレイヤー1-5番（インデックス0-4）: 左チーム
- プレイヤー6-10番（インデックス5-9）: 右チーム

## チーム勝敗判定

### 勝利条件

**チームスコア積による判定**：
- 各チームの全プレイヤーのスコアを掛け合わせた値（積）が**200以上**になったチームが勝利

```
左チームスコア積 = プレイヤー1のスコア × プレイヤー2のスコア × ... × プレイヤー5のスコア
右チームスコア積 = プレイヤー6のスコア × プレイヤー7のスコア × ... × プレイヤー10のスコア
```

### 敗北条件

以下のいずれかの条件でチームが敗北：
1. 相手チームのスコア積が200以上になった場合
2. 自チームの失格プレイヤーが5人になった場合

## UI表示・操作方法

### スコアボード表示

AQL形式では専用の表示レイアウトが使用されます：

#### チーム表示
- 左右にチームが分かれて表示
- 各チームの上部にチーム名とチームスコア積（または勝敗結果）を表示
- チーム状態によって背景色が変化

#### プレイヤー表示
- 各プレイヤーの正解数が「◯pt」形式で表示
- 失格状態のプレイヤーは休み回数表示（例：「3休」）
- 勝ち抜けプレイヤーは順位表示（例：「1st」）

### 操作方法

基本的な操作方法については、[共通操作方法](../common-operations.md)を参照してください。

#### AQL形式固有の特徴

- **自動復活**: 相手チームの誤答時に失格プレイヤーが自動的に復活
- **チーム表示**: リアルタイムでチームスコア積と勝敗状況を表示
- **10人固定**: 必ず10人でゲームを開始する必要がある

## 実装詳細

### 関連ファイル

- **スコア計算ロジック**: `src/utils/computeScore/aql.ts`
- **ゲーム設定定義**: `src/utils/rules.ts` (193-204行目)
- **型定義**: `src/utils/types.ts`
- **専用UIコンポーネント**: `src/app/(board)/games/[game_id]/board/_components/AQL/AQL.tsx`
- **プレイヤー表示**: `src/app/(board)/games/[game_id]/board/_components/AQLPlayer/AQLPlayer.tsx`

### UI コンポーネント

AQL形式では他の形式と異なる専用UIが使用されます：

- **メインボード**: `AQL.tsx` - チーム分け表示とスコア積計算
- **プレイヤー表示**: `AQLPlayer.tsx` - 個別プレイヤーの状態表示
- **チーム状態管理**: リアルタイムでの勝敗判定と表示切り替え

### 計算アルゴリズム

```typescript
// 正解時
case "correct":
  return {
    ...playerState,
    correct: playerState.correct + 1,
    score: playerState.score + 1,  // スコア = 正解数 + 1
    last_correct: qn,
  };

// 誤答時（1回目）
case "wrong":
  if (playerState.reach_state !== "lose") {
    return {
      ...playerState,
      wrong: newWrong,
      score: 1,  // スコアを1にリセット
      reach_state: "lose",  // 誤答リーチ状態
    };
  }

// 誤答時（2回目・失格確定）
  else {
    return {
      ...playerState,
      wrong: newWrong,
      score: 1,
      state: "lose",
      is_incapacity: true,  // 失格フラグ
    };
  }

// 復活処理（相手チーム誤答時）
if (相手チームプレイヤーが誤答 && 自分が失格状態) {
  return {
    ...playerState,
    score: 1,
    state: "playing",
    reach_state: "playing",
    is_incapacity: false,  // 復活
  };
}
```

### チーム勝敗判定アルゴリズム

```typescript
// チームスコア積の計算
const leftTeamScore = playerScoreList.slice(0, 5).reduce((acc, cur) => {
  return acc * cur.score.score;
}, 1);

const rightTeamScore = playerScoreList.slice(5, 10).reduce((acc, cur) => {
  return acc * cur.score.score;
}, 1);

// 勝敗判定
const leftTeamState = 
  leftTeamScore >= 200 ? "win" :
  (rightTeamScore >= 200 || 失格者5人) ? "lose" : "playing";
```

## 使用場面

AQL形式は以下のような場面で活用できます：

- Academic Quiz League形式の大会・練習
- チーム戦略が重要な競技クイズ
- 個人技術とチーム連携の両方が求められる大会
- 復活要素のあるスリリングな競技設計

## 注意事項・制約

### 必須制約
- **プレイヤー数**: 必ず10人で実施（それ以外の人数では正常に動作しない）
- **チーム固定**: プレイヤー1-5番と6-10番の固定チーム分け

### 実装上の課題
- **lose_point未設定**: 現在の実装では休み表示の計算にlose_pointを参照しているが、AQL形式のルール設定では定義されていない
- **設定の不整合**: この問題により休み期間の表示が正しく動作しない可能性がある

### 運用上の注意
- 復活システムにより戦況が大きく変わる可能性があるため、プレイヤーへの事前説明が重要
- チームスコア積による勝利条件は他の形式と大きく異なるため、理解の確認が必要
- 失格者が多数出た場合のゲーム進行についても事前に確認することを推奨

## 戦略的要素

AQL形式特有の戦略的要素：

1. **復活を狙った誤答**: 味方の復活のために意図的に誤答する戦略
2. **スコア積の最適化**: 全員の正解数を均等に上げる重要性
3. **リスク管理**: 誤答リーチ状態での慎重な解答判断
4. **チーム連携**: 相手チームの状況を見極めた戦術変更

これらの要素により、単純な早押し技術だけでなく、高度なチーム戦略が求められる形式となっています。