# アタックサバイバル形式仕様書

## 概要

アタックサバイバル形式（アタックサバイバル）は、競技クイズにおいて他のプレイヤーとの相互作用によってスコアが変動する独特なルール形式です。プレイヤーは各自一定のポイントを持った状態でゲームを開始し、正解すると他の全プレイヤーのポイントを減らすことができ、誤答すると自分のポイントが減ります。ポイントが0になったプレイヤーは失格となり、最終的に生き残ったプレイヤーが勝ち抜けとなる、まさに「サバイバル」形式の競技です。

## 基本ルール

### 勝利条件

- 生き残りプレイヤー数が設定人数以下になったとき、残ったプレイヤーが勝ち抜け
- デフォルト設定：3人まで生き残ったときに勝ち抜け

### 失格条件

- スコアが0点になると失格

### スコア変動システム

- **初期スコア**：15点（設定可能）
- **自分が正解時**：自分のスコア+0点、他の全プレイヤーのスコア-1点
- **自分が誤答時**：自分のスコア-2点、他のプレイヤーのスコア+0点

### リーチ状態

- 次の誤答で失格になる状態でリーチ表示
- 他プレイヤーの正解による影響でも失格リーチ判定

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "attacksurvival",
  win_point: number,        // 初期スコア（デフォルト：15）
  win_through: number,      // 勝ち抜け人数（デフォルト：3）
  correct_me: number,       // 自分正解時の自分へのスコア変動（デフォルト：0）
  wrong_me: number,         // 自分誤答時の自分へのスコア変動（デフォルト：-2）
  correct_other: number,    // 自分正解時の他者へのスコア変動（デフォルト：-1）
  wrong_other: number,      // 自分誤答時の他者へのスコア変動（デフォルト：0）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | attacksurvival形式での扱い | 説明                                           |
| -------------------- | -------- | ---------------------- | -------------------------- | ---------------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用       | ゲーム内でのプレイヤー識別子                   |
| `name`               | `string` | マスターデータより継承 | 表示に使用                 | ゲーム内での表示名                             |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない               | ゲーム開始時の正解数                           |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない               | ゲーム開始時の誤答数                           |
| `base_correct_point` | `number` | 通常15                 | 初期スコアとして使用       | ゲーム開始時のスコア                           |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない               | 誤答時の基本点数（attacksurvivalでは未使用）   |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値      | 正解時           | 誤答時       | 他者の行動による変動 | 説明                     |
| --------------- | --------- | ----------- | ---------------- | ------------ | -------------------- | ------------------------ |
| `score`         | `number`  | `初期値`    | `+correct_me`    | `+wrong_me`  | 相手の行動に応じて変動 | 現在のスコア             |
| `correct`       | `number`  | `0`         | `+1`             | 変更なし     | 変更なし             | 現在の正解数             |
| `wrong`         | `number`  | `0`         | 変更なし         | `+1`         | 変更なし             | 現在の誤答数             |
| `last_correct`  | `number`  | `-1`        | 問題番号         | 変更なし     | 変更なし             | 最後に正解した問題番号   |
| `last_wrong`    | `number`  | `-1`        | 変更なし         | 問題番号     | 変更なし             | 最後に誤答した問題番号   |
| `state`         | `string`  | `"playing"` | 条件により`"win"` | 失格時`"lose"` | 失格時`"lose"`       | プレイヤーの状態         |
| `reach_state`   | `string`  | `"playing"` | 変更なし         | リーチ時`"lose"` | リーチ時`"lose"`     | リーチ状態               |
| `text`          | `string`  | 動的生成    | 動的更新         | 動的更新     | 動的更新             | 表示テキスト             |
| `order`         | `number`  | `0`         | 勝利時に順位設定 | 変更なし     | 変更なし             | 勝ち抜け順位             |

### 計算ロジック

#### 解答者本人への処理

##### 正答時の処理

1. プレイヤーの正解数を1増加
2. プレイヤーのスコアに`correct_me`を加算
3. 次の誤答で失格になる場合、失格リーチ状態に設定

```typescript
// 自分が正解した場合
const newScore = playerState.score + game.correct_me;
if (newScore + game.wrong_me <= 0) {
  // 次の誤答で失格になる場合
  reach_state = "lose";
}
```

##### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. プレイヤーのスコアに`wrong_me`を加算
3. スコアが0以下になった場合、失格状態に設定
4. 次の誤答で失格になる場合、失格リーチ状態に設定

```typescript
// 自分が誤答した場合
const newScore = playerState.score + game.wrong_me;
if (newScore <= 0) {
  // 失格
  state = "lose";
} else if (newScore + game.wrong_me <= 0) {
  // 次の誤答で失格リーチ
  reach_state = "lose";
}
```

#### 他のプレイヤーへの処理

##### 他者の正解による影響

1. 自分のスコアに`correct_other`を加算
2. スコアが0以下になった場合、失格状態に設定
3. 失格リーチ条件を満たす場合、失格リーチ状態に設定

```typescript
// 他者が正解した場合
const newScore = playerState.score + game.correct_other;
if (newScore <= 0) {
  // 失格
  state = "lose";
} else if (
  newScore + game.correct_other <= 0 ||
  newScore + game.wrong_me <= 0
) {
  // 失格リーチ
  reach_state = "lose";
}
```

##### 他者の誤答による影響

1. 自分のスコアに`wrong_other`を加算
2. 失格・リーチ判定を実行

#### 勝利判定

生き残っているプレイヤー数が`win_through`以下になった場合、残ったプレイヤーが勝ち抜けとなります。

```typescript
const playingPlayers = playersState.filter((p) => p.state === "playing");
const currentState =
  game.win_through &&
  game.win_through >= playingPlayers.length &&
  playerState.state === "playing"
    ? "win"
    : playerState.state;
```

### UI表示仕様

#### スコアボード表示

2行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時："LOSE"
  - 通常時：現在のスコア（例："12pt"）

- **2行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）

#### 色分け

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色
  - 失格時：失格状態の色

- **操作ボタン**：正解は赤色、誤答は青色

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← スコア・状態表示
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 初期スコア（win_point）

- 範囲：1-1000
- デフォルト：15
- 説明：ゲーム開始時の各プレイヤーのスコア

#### 勝ち抜け人数（win_through）

- 範囲：1-100
- デフォルト：3
- 説明：この人数まで生き残ったときに勝ち抜け

#### 自分正解時の自分への影響（correct_me）

- 範囲：-1000-1000
- デフォルト：0
- 説明：自分が正解したときの自分のスコア変動

#### 自分誤答時の自分への影響（wrong_me）

- 範囲：-1000-1000
- デフォルト：-2
- 説明：自分が誤答したときの自分のスコア変動

#### 自分正解時の他者への影響（correct_other）

- 範囲：-1000-1000
- デフォルト：-1
- 説明：自分が正解したときの他プレイヤーのスコア変動

#### 自分誤答時の他者への影響（wrong_other）

- 範囲：-1000-1000
- デフォルト：0
- 説明：自分が誤答したときの他プレイヤーのスコア変動

## 戦略的要素

### 相互作用システム

アタックサバイバル形式の最大の特徴は、プレイヤー間の相互作用です：

1. **攻撃性の重要性**：正解により他者を攻撃できる
2. **防御の必要性**：誤答による自分へのダメージを避ける
3. **集団戦略**：全体の生き残り状況を考慮した判断

### 数学的特性

#### スコア変動の分析

**デフォルト設定での影響**：

| 行動 | 自分への影響 | 他者への影響 | 全体への影響 |
|------|--------------|--------------|--------------|
| 正解 | +0点         | -1点/人      | 全体でマイナス |
| 誤答 | -2点         | +0点/人      | 自分のみマイナス |

#### 生存可能性の計算

- **初期スコア15点の場合**
- 誤答のみで失格：7.5回（実質7回）
- 他者の正解のみで失格：15回
- 混合パターンでの計算が重要

### プレイヤーの戦略

1. **攻撃型戦略**：積極的に正解して他者を削る
2. **守備型戦略**：誤答を避けて生き残りを優先
3. **状況判断型戦略**：残り人数と自分のスコアを考慮
4. **協調型戦略**：特定のプレイヤーと暗黙の協力
5. **カウンター戦略**：強いプレイヤーの攻撃を意識

### 心理的要素

- **攻撃のプレッシャー**：他者への影響を考慮した解答
- **標的意識**：スコアの高いプレイヤーへの注目
- **生存不安**：残り人数と自分の順位への意識
- **終盤の緊張**：勝ち抜け圏内での慎重さ

## 他形式との比較

### nomx形式との違い

- **nomx**：個人の正誤のみでスコア変動
- **attacksurvival**：他者の行動も自分のスコアに影響

### ny形式との違い

- **ny**：固定的なスコア変動
- **attacksurvival**：他者との相互作用による変動

### divide形式との違い

- **divide**：個人の誤答履歴による変動
- **attacksurvival**：リアルタイムな相互作用

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- 全プレイヤーへの影響計算による処理負荷
- リアルタイムな相互作用の効率的な実装
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

### 複雑性の管理

- 各プレイヤーの行動が全体に与える影響の正確な計算
- リーチ状態の多面的な判定（誤答リーチ、他者攻撃リーチ）
- 勝利条件の動的な判定

## 使用例

### 設定例1：標準的なアタックサバイバル設定

- 初期スコア：15点
- 勝ち抜け人数：3人
- 自分正解時：自分+0、他者-1
- 自分誤答時：自分-2、他者+0
- 用途：一般的な早押しクイズ大会

### 設定例2：攻撃重視設定

- 初期スコア：20点
- 勝ち抜け人数：2人
- 自分正解時：自分+1、他者-2
- 自分誤答時：自分-3、他者+0
- 用途：攻撃性を重視した競技

### 設定例3：防御重視設定

- 初期スコア：10点
- 勝ち抜け人数：4人
- 自分正解時：自分+0、他者-1
- 自分誤答時：自分-1、他者+0
- 用途：慎重さを重視した競技

### 設定例4：相互影響強化設定

- 初期スコア：15点
- 勝ち抜け人数：3人
- 自分正解時：自分+1、他者-2
- 自分誤答時：自分-2、他者+1
- 用途：より強い相互作用を求める競技

## 競技性の特徴

### 適用場面

1. **戦略的思考力**：相互作用を考慮した判断能力
2. **状況把握能力**：全体の流れと自分の位置の認識
3. **プレッシャー耐性**：他者への影響を意識した解答
4. **協調・競争バランス**：時には協力、時には競争

### 出題者の配慮

- **問題難易度の調整**：攻撃・防御のバランス
- **進行速度の管理**：戦略的思考の時間確保
- **心理的負荷の配慮**：過度なプレッシャーの回避

## 実装上の課題とTODO

### 現在の制限事項

#### 1. 相互作用の可視化

**現状**：基本的なスコア表示のみ
**改善案**：他者への影響度の明示的表示

#### 2. 戦略ガイダンス

**現状**：数値情報のみ
**改善案**：最適戦略や状況分析の提供

#### 3. 履歴管理

**現状**：リアルタイム情報のみ
**改善案**：攻撃・被攻撃の履歴表示

### 将来的な改善提案

1. **相互作用の視覚化**：攻撃効果の分かりやすい表示
2. **戦略分析機能**：各プレイヤーの行動傾向分析
3. **シミュレーション機能**：仮想的な行動結果の予測
4. **協調度測定**：プレイヤー間の協力・競争関係の分析

この形式は、従来の個人戦中心のクイズ形式とは異なり、他者との相互作用を重視した戦略的な競技形式として設計されています。プレイヤーには知識力だけでなく、状況判断力と戦略的思考力が要求される、高度に発展したルールです。