# Backstream形式仕様書

## 概要

Backstream形式（バックストリーム）は、競技クイズにおいて誤答回数に応じて減点幅が増加する独特なルール形式です。1回の正答で+1点を獲得し、n回目の誤答で-n点という累積的なペナルティシステムにより、10点で勝ち抜け、-10点で失格を目指します。誤答を重ねるほどスコアの下落が激しくなるため、慎重さと正確性が重要となる戦略性の高い形式です。

## 基本ルール

### 勝利条件

- スコアが10点に達することで勝ち抜け
- デフォルト設定：10点で勝ち抜け

### 失格条件

- スコアが-10点以下になると失格
- デフォルト設定：-10点以下で失格

### スコア計算システム

- **初期スコア**：0点
- **正解時**：スコア+1点
- **誤答時**：スコア - 誤答回数

### 誤答時の減点システム

誤答時の減点は、その時点での累積誤答回数によって決定されます：

- **1回目の誤答**：-1点
- **2回目の誤答**：-2点
- **3回目の誤答**：-3点
- **n回目の誤答**：-n点

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "backstream",
  win_point: number,    // 勝ち抜けに必要なスコア（デフォルト：10）
  lose_point: number,   // 失格となるスコア（デフォルト：-10）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | backstream形式での扱い | 説明                                     |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子             |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名                       |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の正解数                     |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の誤答数                     |
| `base_correct_point` | `number` | 通常1                  | 使用されない           | 正解時の基本点数（backstreamでは未使用） |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない           | 誤答時の基本点数（backstreamでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値      | 正解時           | 誤答時         | 説明                       |
| -------------- | -------- | ----------- | ---------------- | -------------- | -------------------------- |
| `score`        | `number` | `0`         | `+1`             | `-誤答回数`    | 現在のスコア               |
| `correct`      | `number` | `0`         | `+1`             | 変更なし       | 現在の正解数               |
| `wrong`        | `number` | `0`         | 変更なし         | `+1`           | 現在の誤答数               |
| `last_correct` | `number` | `-1`        | 問題番号         | 変更なし       | 最後に正解した問題番号     |
| `last_wrong`   | `number` | `-1`        | 変更なし         | 問題番号       | 最後に誤答した問題番号     |
| `state`        | `string` | `"playing"` | 勝利時`"win"`    | 失格時`"lose"` | プレイヤーの状態           |
| `reach_state`  | `string` | `"playing"` | 変更なし         | 変更なし       | リーチ状態（使用されない） |
| `text`         | `string` | 動的生成    | 動的更新         | 動的更新       | 表示テキスト               |
| `order`        | `number` | `0`         | 勝利時に順位設定 | 変更なし       | 勝ち抜け順位               |

### 計算ロジック

#### 正答時の処理

1. プレイヤーのスコアを1増加
2. プレイヤーの正解数を1増加
3. スコアが`win_point`（10点）に到達した場合、勝ち抜け状態に設定

```typescript
const newCorrect = playerState.correct + 1;
const newScoreInCorrectCase = playerState.score + 1;
if (newScoreInCorrectCase >= game.win_point!) {
  // 勝ち抜け
  state = "win";
}
```

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. スコアから新しい誤答回数を減算
3. スコアが`lose_point`（-10点）以下になった場合、失格状態に設定

```typescript
const newWrong = playerState.wrong + 1;
const newScoreInWrongCase = playerState.score - newWrong;
if (newScoreInWrongCase <= game.lose_point!) {
  // 失格
  state = "lose";
}
```

#### 具体的な計算例

**初期状態：0点**

1. 正解 → 0 + 1 = 1点
2. 正解 → 1 + 1 = 2点
3. 誤答（1回目） → 2 - 1 = 1点
4. 正解 → 1 + 1 = 2点
5. 誤答（2回目） → 2 - 2 = 0点
6. 正解 → 0 + 1 = 1点
7. 誤答（3回目） → 1 - 3 = -2点
8. 誤答（4回目） → -2 - 4 = -6点
9. 誤答（5回目） → -6 - 5 = -11点（失格）

### UI表示仕様

#### スコアボード表示

2行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時："LOSE"
  - 通常時：現在のスコア（例："3pt"、"-2pt"）

- **2行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）

#### 色分け

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色
  - 失格時：失格状態の色

- **操作ボタン**：正解は赤色、誤答は青色

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← スコア・状態表示
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：10
- 説明：この点数に達すると勝ち抜け

#### 失格ポイント（lose_point）

- 範囲：-1000-0
- デフォルト：-10
- 説明：この点数以下になると失格

## 戦略的要素

### 累積的誤答ペナルティ

Backstream形式の最大の特徴は、誤答を重ねるほど減点幅が大きくなることです：

1. **初回誤答**：軽微なダメージ（-1点）
2. **2回目誤答**：中程度のダメージ（-2点）
3. **3回目以降**：深刻なダメージ（-3点、-4点、...）

### 数学的特性

#### スコア変動の分析

**累積誤答による減点合計**：

| 誤答回数 | 減点 | 累積減点 |
| -------- | ---- | -------- |
| 1回      | -1   | -1       |
| 2回      | -2   | -3       |
| 3回      | -3   | -6       |
| 4回      | -4   | -10      |
| 5回      | -5   | -15      |

**必要正解数の計算**：

- 勝ち抜けには10点が必要
- n回誤答後は n(n+1)/2 点の減点
- 4回誤答後（-10点）からの勝ち抜けには20回正解が必要

#### 危険領域の分析

- **0-2回誤答**：比較的安全な段階
- **3回誤答**：要注意段階（累積-6点）
- **4回誤答**：危険段階（累積-10点、失格直前）

### プレイヤーの戦略

1. **慎重型戦略**：誤答を最小限に抑える確実性重視
2. **計算型戦略**：現在のスコアと許容誤答回数を計算
3. **積極型戦略**：早期の正解蓄積でリスクヘッジ
4. **復活型戦略**：マイナススコアからの長期戦略

### 心理的要素

- **累積プレッシャー**：誤答を重ねるほど高まる緊張感
- **リスク意識**：次の誤答のダメージを常時把握
- **諦めへの誘惑**：深いマイナスからの復活の困難さ

## 他形式との比較

### ny形式との違い

- **ny**：誤答で-1点（固定減点）
- **backstream**：誤答で-n点（累積的減点）

### divide形式との違い

- **divide**：誤答で除算（乗法的減少）
- **backstream**：誤答で累積減点（加法的減少）

### nupdown形式との違い

- **nupdown**：誤答でスコアリセット（一律ペナルティ）
- **backstream**：誤答で累積減点（段階的ペナルティ）

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- 単純な減算処理による効率的なスコア計算
- UI更新は状態変更時のみ実行され、不要な再描画を抑制
- 負のスコア表示への対応

### 数値範囲の考慮

- 負のスコア値の適切な処理
- 大きな減点値での計算精度の確保
- 失格判定の正確な実装

## 使用例

### 設定例1：標準的なBackstream設定

- 勝ち抜けポイント：10
- 失格ポイント：-10
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜けポイント：7
- 失格ポイント：-7
- 用途：時間制限のある競技

### 設定例3：寛容設定

- 勝ち抜けポイント：10
- 失格ポイント：-15
- 用途：学習目的の競技

### 設定例4：厳格設定

- 勝ち抜けポイント：15
- 失格ポイント：-8
- 用途：高精度を要求する競技

## 競技性の特徴

### 適用場面

1. **累積リスク管理**：段階的に増加するリスクへの対応力
2. **計算能力の評価**：数学的なリスク予測能力
3. **復活力の測定**：不利な状況からの回復能力
4. **プレッシャー耐性**：累積的緊張への対応力

### 出題者の配慮

- **問題難易度の調整**：段階的なリスク増加への配慮
- **復活機会の提供**：マイナススコアからの逆転可能性
- **時間配分**：計算時間を含めた進行管理

## 実装上の課題とTODO

### 現在の制限事項

#### 1. リスク可視化の強化

**現状**：基本的なスコア表示のみ
**改善案**：次回誤答時の減点値の明示的表示

#### 2. 復活可能性の表示

**現状**：現在のスコアのみ表示
**改善案**：勝ち抜けまでに必要な正解数の表示

#### 3. 戦略ガイダンス

**現状**：数値情報のみ
**改善案**：リスクレベルや最適戦略のアドバイス

### 将来的な改善提案

1. **累積減点の視覚化**：減点履歴の分かりやすい表示
2. **予測機能**：次回誤答時のスコア予測表示
3. **統計機能**：各誤答回数での復活率の分析
4. **学習支援機能**：最適解答タイミングの提案

この形式は、誤答に対して累積的かつ段階的なペナルティを設けることで、プレイヤーに継続的なリスク管理と数学的思考を要求する高度な戦略性を持つルールです。失格の可能性と復活の可能性を両立させることで、最後まで緊張感を維持できる競技形式として設計されています。
