# Divide形式仕様書

## 概要

Divide形式（ディバイド）は、競技クイズにおいて誤答回数に応じてスコアの割る数が増加する独特なルール形式です。プレイヤーは初期値として一定のポイントを持った状態でゲームを開始し、正解するごとに一定のポイントが加算されます。しかし誤答するたびに現在のスコアが誤答回数で割られ、誤答を重ねるほどスコアの減少が大きくなる設計となっています。100点に達すると勝ち抜けとなり、失格条件は存在しないため、すべてのプレイヤーが最後まで競技に参加できます。

## 基本ルール

### 勝利条件

- スコアが100点に達することで勝ち抜け
- デフォルト設定：100点で勝ち抜け

### 失格条件

- 失格条件は存在しない
- スコアが0点になっても競技続行可能

### スコア計算システム

- **初期スコア**：10点
- **正解時**：スコア + 10点
- **誤答時**：現在のスコア ÷ 誤答回数（小数点以下切り捨て）

### 誤答時の除算ルール

- 1回目の誤答：スコア ÷ 1
- 2回目の誤答：スコア ÷ 2
- 3回目の誤答：スコア ÷ 3
- N回目の誤答：スコア ÷ N

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "divide",
  win_point: number,    // 勝ち抜けに必要なスコア（デフォルト：100）
  correct_me: number,   // 正解時の加算ポイント（デフォルト：10）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | divide形式での扱い   | 説明                                 |
| -------------------- | -------- | ---------------------- | -------------------- | ------------------------------------ |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用 | ゲーム内でのプレイヤー識別子         |
| `name`               | `string` | マスターデータより継承 | 表示に使用           | ゲーム内での表示名                   |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない         | ゲーム開始時の正解数                 |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない         | ゲーム開始時の誤答数                 |
| `base_correct_point` | `number` | 通常10                 | 使用されない         | 正解時の基本点数（divideでは未使用） |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない         | 誤答時の基本点数（divideでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値      | 正解時           | 誤答時      | 説明                   |
| -------------- | -------- | ----------- | ---------------- | ----------- | ---------------------- |
| `score`        | `number` | `10`        | `+correct_me`    | `÷誤答回数` | 現在のスコア           |
| `correct`      | `number` | `0`         | `+1`             | 変更なし    | 現在の正解数           |
| `wrong`        | `number` | `0`         | 変更なし         | `+1`        | 現在の誤答数           |
| `last_correct` | `number` | `-1`        | 問題番号         | 変更なし    | 最後に正解した問題番号 |
| `last_wrong`   | `number` | `-1`        | 変更なし         | 問題番号    | 最後に誤答した問題番号 |
| `state`        | `string` | `"playing"` | 勝利時`"win"`    | 変更なし    | プレイヤーの状態       |
| `reach_state`  | `string` | `"playing"` | リーチ時`"win"`  | 変更なし    | リーチ状態             |
| `text`         | `string` | 動的生成    | 動的更新         | 動的更新    | 表示テキスト           |
| `order`        | `number` | `0`         | 勝利時に順位設定 | 変更なし    | 勝ち抜け順位           |

### 計算ロジック

#### 正答時の処理

1. プレイヤーのスコアに`correct_me`を加算
2. プレイヤーの正解数を1増加
3. スコアが`win_point`に到達した場合、勝ち抜け状態に設定
4. スコアが`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定

```typescript
const newScoreWhenCorrect = playerState.score + game.correct_me;
if (newScoreWhenCorrect >= game.win_point!) {
  // 勝ち抜け
  state = "win";
} else if (newScoreWhenCorrect + 1 === game.win_point!) {
  // 勝ち抜けリーチ
  reach_state = "win";
}
```

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. 現在のスコアを新しい誤答回数で割る（小数点以下切り捨て）
3. 失格条件は存在しないため、競技続行

```typescript
const newScoreWhenWrong = Math.floor(
  playerState.score / (playerState.wrong + 1)
);
```

#### 具体的な計算例

**初期状態：10点**

1. 正解 → 10 + 10 = 20点
2. 正解 → 20 + 10 = 30点
3. 誤答（1回目） → Math.floor(30 ÷ 1) = 30点
4. 正解 → 30 + 10 = 40点
5. 誤答（2回目） → Math.floor(40 ÷ 2) = 20点
6. 正解 → 20 + 10 = 30点
7. 誤答（3回目） → Math.floor(30 ÷ 3) = 10点

### UI表示仕様

#### スコアボード表示

2行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 通常時：現在のスコア（例："45pt"）

- **2行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）

#### 色分け

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色

- **操作ボタン**：正解は赤色、誤答は青色

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← スコア・状態表示
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：100
- 説明：この点数に達すると勝ち抜け

#### 正解時加算ポイント（correct_me）

- 範囲：1-1000
- デフォルト：10
- 説明：正解時に加算されるポイント

## 戦略的要素

### 誤答の累積効果

Divide形式の最大の特徴は、誤答の累積効果です：

1. **初回誤答**：ダメージが軽微（÷1 = 変化なし）
2. **2回目誤答**：中程度のダメージ（÷2 = 半減）
3. **3回目以降**：深刻なダメージ（÷3, ÷4, ...）

### 数学的特性

#### スコア減少パターン（誤答時）

| 誤答回数 | 除算値 | 60点の場合 | 30点の場合 | 15点の場合 |
| -------- | ------ | ---------- | ---------- | ---------- |
| 1回目    | ÷1     | 60点       | 30点       | 15点       |
| 2回目    | ÷2     | 30点       | 15点       | 7点        |
| 3回目    | ÷3     | 20点       | 10点       | 5点        |
| 4回目    | ÷4     | 15点       | 7点        | 3点        |
| 5回目    | ÷5     | 12点       | 6点        | 3点        |

#### 回復可能性の分析

誤答後の回復に必要な正解数：

- **2回誤答後（スコア半減）**：回復に2回正解が必要
- **3回誤答後（スコア1/3）**：回復に3回正解が必要
- **4回誤答後（スコア1/4）**：回復に4回正解が必要

### プレイヤーの戦略

1. **慎重型戦略**：誤答を避けて確実な問題のみ解答
2. **積極型戦略**：早期に高スコアを狙い、誤答リスクを取る
3. **計算型戦略**：現在のスコアと誤答回数を考慮した戦略的判断
4. **復活型戦略**：多誤答後でも諦めず、長期戦で巻き返しを狙う

### 心理的要素

- **誤答への恐怖心**：高スコア時ほど誤答のダメージが深刻
- **諦めの防止**：失格がないため、最後まで参加可能
- **リスク評価**：現在のスコアと誤答回数による戦略変更

## 他形式との比較

### ny形式との違い

- **ny**：誤答で-1点（線形減少）
- **divide**：誤答で除算（非線形減少、累積効果あり）

### nupdown形式との違い

- **nupdown**：誤答でスコアリセット（一律0点）
- **divide**：誤答で除算（スコアに応じた減少）

### nomx形式との違い

- **nomx**：誤答で失格（永続的ペナルティ）
- **divide**：誤答でスコア減少（回復可能、失格なし）

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- 除算と切り捨て処理による効率的なスコア計算
- UI更新は状態変更時のみ実行され、不要な再描画を抑制
- 失格判定が不要なため、シンプルな状態管理

### 数値精度の考慮

- `Math.floor()`による小数点以下切り捨て処理
- 整数演算による精度の確保
- 0点到達後の動作の安定性

## 使用例

### 設定例1：標準的な設定

- 勝ち抜けポイント：100
- 正解時加算ポイント：10
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜けポイント：50
- 正解時加算ポイント：10
- 用途：時間制限のある競技

### 設定例3：長期戦設定

- 勝ち抜けポイント：200
- 正解時加算ポイント：15
- 用途：実力重視の長時間競技

### 設定例4：細かい刻み設定

- 勝ち抜けポイント：100
- 正解時加算ポイント：5
- 用途：より多くの問題数での競技

## 競技性の特徴

### 適用場面

1. **回復力の評価**：不利な状況からの巻き返し能力
2. **リスク管理能力**：現在のスコアを考慮した判断力
3. **持続力の測定**：失格がないため、最後まで競技参加
4. **計算能力**：スコア変動の予測と戦略立案

### 出題者の配慮

- **問題数の調整**：十分な逆転機会の提供
- **難易度バランス**：確実に答えられる問題の混在
- **時間管理**：長期戦に対応した進行

## 実装上の課題とTODO

### 現在の制限事項

#### 1. 初期スコア設定の柔軟性

**現状**：固定的な初期値10点
**改善案**：ゲーム設定での初期スコア調整機能

#### 2. 除算結果の表示

**現状**：最終結果のみ表示
**改善案**：除算計算過程の可視化

#### 3. 戦略ガイダンス

**現状**：基本的なスコア表示のみ
**改善案**：現在の誤答回数に応じたリスク表示

### 将来的な改善提案

1. **統計情報の追加**：平均スコア、誤答率の表示
2. **計算履歴の表示**：スコア変動の詳細記録
3. **戦略アドバイス**：最適解答タイミングの提案
4. **視覚的効果**：除算時のアニメーション表示

この形式は、誤答に対して累積的なペナルティを設けながらも、失格による脱落を防ぐことで、すべてのプレイヤーが最後まで競技に参加できる包括的なルールとして設計されています。誤答の累積効果により、慎重さと積極性のバランスが重要となる戦略性の高い競技形式です。

## ユニットテスト観点

ユニットテストでは`getInitialPlayersStateForOnline`がdivide形式に対してゲームオプションの`correct_me`値を初期スコアとして採用し、プレイヤーごとの`initialScore`値は正解数・誤答数にのみ反映されることを確かめる。初期スコアの異なる複数プレイヤーを用意し、初期状態でのスコア差が正しく反映されるか、さらに同値スコア時に正解数や誤答数の比較まで行われる並び順判定が期待どおりに動くかを検証する。これにより、誤答による除算処理に入る前段階の前提条件を保証する。
