# エンドレスチャンス形式仕様書

## 概要

エンドレスチャンス形式は、競技クイズにおいて同じ問題に対して正答が出るまで複数人が回答できるルール形式です。正解者が出るかスルーになるまで次の問題に進まず、すべてのプレイヤーが各問題に挑戦する機会を持てる設計となっています。誤答に対するペナルティは設定により失格または休みを選択できます。

## 基本ルール

### 問題進行の仕組み

- 1つの問題に対して、正解が出るまで複数のプレイヤーが順次回答可能
- 誰かが正解するか、出題者がスルーを選択するまで同じ問題を継続
- まだ回答していないプレイヤーには回答権が残る

### 勝利条件

- N回正解することで勝ち抜け
- デフォルト設定：7回正解で勝ち抜け

### ペナルティ条件

誤答時のペナルティは設定により2パターンから選択可能：

#### パターン1：失格制（休み機能オフ）
- M回誤答すると失格
- デフォルト設定：3回誤答で失格

#### パターン2：休み制（休み機能オン）
- M回誤答するとM回「休み」状態になり、解答権を失う
- デフォルト設定：3問休み
- 何回誤答しても最終的な失格はない

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "endless-chance",
  win_point: number,    // 勝ち抜けに必要な正解数（デフォルト：7）
  lose_point: number,   // 失格誤答数または休み回数（デフォルト：3）
  options: {
    use_r: boolean      // 休み機能の有効/無効（デフォルト：false）
  }
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | エンドレスチャンス形式での扱い    | 説明                                      |
| -------------------- | -------- | ---------------------- | --------------------------------- | ----------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用              | ゲーム内でのプレイヤー識別子              |
| `name`               | `string` | マスターデータより継承 | 表示に使用                        | ゲーム内での表示名                        |
| `initial_correct`    | `number` | 通常0（調整可能）      | スコア初期値として使用            | ゲーム開始時の正解数                      |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 計算に影響あり                    | ゲーム開始時の誤答数                      |
| `base_correct_point` | `number` | 通常10                 | 使用されない                      | 正解時の基本点数（エンドレスチャンスでは未使用） |
| `base_wrong_point`   | `number` | 通常-10                | 使用されない                      | 誤答時の基本点数（エンドレスチャンスでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値            | 正解時           | 誤答時   | 休み明け時      | 説明                            |
| --------------- | --------- | ----------------- | ---------------- | -------- | --------------- | ------------------------------- |
| `score`         | `number`  | `initial_correct` | 変更なし         | 変更なし | 変更なし        | エンドレスチャンス形式では使用されない |
| `correct`       | `number`  | `initial_correct` | `+1`             | 変更なし | 変更なし        | 現在の正解数                    |
| `wrong`         | `number`  | `initial_wrong`   | 変更なし         | `+1`     | 変更なし        | 現在の誤答数                    |
| `last_correct`  | `number`  | `-10`             | 問題番号         | 変更なし | 変更なし        | 最後に正解した問題番号          |
| `last_wrong`    | `number`  | `-10`             | 変更なし         | 問題番号 | 変更なし        | 最後に誤答した問題番号          |
| `is_incapacity` | `boolean` | `false`           | 変更なし         | 休み機能時`true` | `false` | 休み状態フラグ                  |
| `state`         | `string`  | `"playing"`       | 勝利時`"win"`    | 失格時`"lose"` | `"playing"` | プレイヤーの状態                |
| `reach_state`   | `string`  | `"normal"`        | リーチ時`"win"`  | 変更なし | `"normal"`      | リーチ状態                      |
| `text`          | `string`  | 動的生成          | 動的更新         | 動的更新 | 動的更新        | 表示テキスト                    |
| `order`         | `number`  | `0`               | 勝利時に順位設定 | 変更なし | 変更なし        | 勝ち抜け順位                    |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. 正解数が`win_point`に到達した場合、勝ち抜け状態に設定
3. 正解数が`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定
4. 問題を次に進める

#### 誤答時の処理

##### 休み機能無効時（use_r: false）
1. プレイヤーの誤答数を1増加
2. 誤答数が`lose_point`に到達した場合、失格状態に設定
3. `last_wrong`に現在の問題番号を記録
4. 複数誤答ログ（`multiple_wrong`）に誤答プレイヤーを追加

##### 休み機能有効時（use_r: true）
1. プレイヤーの誤答数を1増加
2. `is_incapacity`を`true`に設定（休み状態）
3. `last_wrong`に現在の問題番号を記録
4. 複数誤答ログ（`multiple_wrong`）に誤答プレイヤーを追加

#### スルー時の処理

1. 問題を次に進める
2. プレイヤーの状態変更なし

#### 複数誤答ログの仕組み

エンドレスチャンス形式では、1つの問題で複数人が誤答する可能性があるため、特殊な`multiple_wrong`ログを使用：

```typescript
// 複数誤答ログの構造例
{
  id: "log_id",
  game_id: "game_id", 
  variant: "multiple_wrong",
  player_id_list: ["player1_id", "player2_id", "player3_id"],
  timestamp: "2024-01-01T00:00:00.000Z"
}
```

#### 休み状態の回復判定（休み機能有効時）

```typescript
// 休み状態から回復する条件
const isRecovered =
  game.lose_point! < gameLogList.length - playerState.last_wrong;
```

休み状態からの回復は、最後に誤答した問題番号からの経過問題数が`lose_point`以上になると解答権が回復します。

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時：「失格」
  - 休み中：残り休み回数（例：「2休」）
  - 通常時：現在のポイント数（例：「5pt」）

- **操作ボタン**
  - 正解ボタン（赤色）：休み中はグレーアウトして無効化
  - 誤答ボタン（青色）：休み中はグレーアウトして無効化、複数誤答ログの管理機能付き

- **色分け**
  - 正解ボタン：通常時は赤色、休み・失格時は無効化
  - 誤答ボタン：通常時は青色、休み・失格時は無効化
  - 勝ち抜け時：金色

#### レイアウト構成

2行構成でプレイヤー情報を表示：

- 上段：メイン状態表示
- 下段：正解・誤答操作ボタン

#### 特殊な誤答ボタン機能

誤答ボタンは複数誤答ログの管理機能を持ち：

- 最後のログが`multiple_wrong`の場合、プレイヤーをログに追加/削除
- 同一問題で複数人の誤答を効率的に管理

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：7
- 説明：この回数正解すると勝ち抜け

#### 失格誤答数/休み回数（lose_point）

- 範囲：1-100
- デフォルト：3
- 説明：
  - 休み機能無効時：この回数誤答すると失格
  - 休み機能有効時：誤答時にこの問題数だけ休み状態になる

#### 休み機能（use_r）

- デフォルト：false（失格制）
- true：誤答時に休み制を採用
- false：誤答時に失格制を採用

## 他形式との比較

### nomr形式との違い

- **nomr**：毎回固定のM問休み、失格なし
- **エンドレスチャンス**：設定により失格制/休み制を選択可能、同一問題で複数人回答可能

### nomx形式との違い

- **nomx**：誤答で即失格、問題は順次進行
- **エンドレスチャンス**：設定により失格制/休み制選択、正解まで同一問題継続

### 通常の早押し形式との違い

- **通常形式**：1問1答、誤答後は次の問題
- **エンドレスチャンス**：1問多答、正解まで継続

## 戦略的要素

### プレイヤーの戦略

1. **順番戦略**：他プレイヤーの解答を見てから判断するか、積極的に先制するか
2. **リスク管理**：失格/休みリスクと勝ち抜けチャンスのバランス
3. **問題選択**：確実に答えられる問題での積極性調整

### 出題者の配慮

- スルーのタイミング判断
- 問題の難易度調整（全員誤答を避ける）
- 休み中プレイヤーが多い場合の進行管理

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 最終正解問題番号の早さ
3. 現在のスコア
4. 正解数
5. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- 複数誤答ログによる効率的な同一問題管理
- 休み機能の有無による条件分岐の最適化
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

## 使用例

### 設定例1：標準的な失格制

- 勝ち抜けポイント：7
- 失格誤答数：3
- 休み機能：無効
- 用途：一般的なエンドレスチャンス早押し

### 設定例2：休み制採用

- 勝ち抜けポイント：7
- 休み回数：3
- 休み機能：有効
- 用途：全員参加型の競技

### 設定例3：短期戦設定

- 勝ち抜けポイント：5
- 失格誤答数：2
- 休み機能：無効
- 用途：時間制限のある競技

### 設定例4：長期戦設定

- 勝ち抜けポイント：10
- 休み回数：4
- 休み機能：有効
- 用途：実力差を明確にしたい競技

この形式は、従来の早押しクイズに「全員に回答機会を与える」という要素を加えたバランスの取れたルールとして設計されています。

## 実装上の課題とTODO

### 現在未実装の機能

#### 1. ゲーム終了条件の自動判定

**現状**：指定人数以下になった場合の自動勝ち抜け判定が未実装
**提案**：残りプレイヤー数が勝ち抜け枠以下になった時点での自動終了機能

#### 2. 複数誤答ログの整合性チェック

**現状**：複数誤答ログの操作時のバリデーションが不十分
**対策**：既に誤答したプレイヤーの重複追加防止、ログの整合性確保

### バリデーション強化が必要な項目

#### 1. 休み中プレイヤーの誤答防止

**現状**：クライアント側でボタンを無効化しているが、サーバーサイドでのバリデーションが不十分
**対策**：休み状態のプレイヤーからの解答操作を受け付けないよう、バックエンドレベルでのバリデーションを追加

#### 2. 勝ち抜け後プレイヤーの操作制限

**現状**：勝ち抜け後も操作が可能な状態
**対策**：勝ち抜け状態（`state: "win"`）のプレイヤーからの操作を無効化するバリデーション

### 将来的な改善提案

1. **ゲーム進行の自動化**：設定した条件での自動ゲーム終了
2. **統計機能の拡張**：問題別の正答率、プレイヤー別の回答パターン分析
3. **UI/UXの改善**：複数誤答の視覚的表現、休み状態の分かりやすい表示
4. **テストカバレッジの向上**：エンドレスチャンス固有のロジックに対するテストケース追加