# NbyN形式仕様書

## 概要

NbyN形式（エヌバイエヌ）は、競技クイズにおいて正答数と誤答数の関係性を利用した独特なスコア計算システムを持つルール形式です。プレイヤーのスコアは「正答数 × (N - 誤答数)」で計算され、N²（Nの2乗）に達することで勝ち抜けを目指します。誤答が増えるほどスコアの上昇率が下がり、戦略的な思考が重要となる形式です。

## 基本ルール

### 勝利条件

- スコアがN²に達することで勝ち抜け
- デフォルト設定：25点（5×5）で勝ち抜け

### 失格条件

- 誤答数がN回に達すると失格
- デフォルト設定：5回誤答で失格

### スコア計算式

```
スコア = 正答数 × (N - 誤答数)
```

- **N**：設定可能な基準値（デフォルト：5）
- 正答数が増えるとスコアが上昇
- 誤答数が増えるとスコアの上昇率が低下

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "nbyn",
  win_point: number,    // 基準値N（デフォルト：5）
  lose_point: number,   // 失格となる誤答数（デフォルト：5）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | nbyn形式での扱い     | 説明                               |
| -------------------- | -------- | ---------------------- | -------------------- | ---------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用 | ゲーム内でのプレイヤー識別子       |
| `name`               | `string` | マスターデータより継承 | 表示に使用           | ゲーム内での表示名                 |
| `initial_correct`    | `number` | 通常0（調整可能）      | スコア計算の初期値   | ゲーム開始時の正解数               |
| `initial_wrong`      | `number` | 通常0（調整可能）      | スコア計算の初期値   | ゲーム開始時の誤答数               |
| `base_correct_point` | `number` | 通常1                  | 使用されない         | 正解時の基本点数（nbynでは未使用） |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない         | 誤答時の基本点数（nbynでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値            | 正解時           | 誤答時           | 説明                   |
| -------------- | -------- | ----------------- | ---------------- | ---------------- | ---------------------- |
| `score`        | `number` | `0`               | 計算式により更新 | 計算式により更新 | 現在のスコア           |
| `correct`      | `number` | `initial_correct` | `+1`             | 変更なし         | 現在の正解数           |
| `wrong`        | `number` | `initial_wrong`   | 変更なし         | `+1`             | 現在の誤答数           |
| `last_correct` | `number` | `-1`              | 問題番号         | 変更なし         | 最後に正解した問題番号 |
| `last_wrong`   | `number` | `-1`              | 変更なし         | 問題番号         | 最後に誤答した問題番号 |
| `state`        | `string` | `"playing"`       | 勝利時`"win"`    | 失格時`"lose"`   | プレイヤーの状態       |
| `reach_state`  | `string` | `"normal"`        | リーチ時`"win"`  | 変更なし         | リーチ状態             |
| `text`         | `string` | 動的生成          | 動的更新         | 動的更新         | 表示テキスト           |
| `order`        | `number` | `0`               | 勝利時に順位設定 | 変更なし         | 勝ち抜け順位           |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. スコアを `正答数 × (win_point - 誤答数)` で再計算
3. スコアが `win_point²` に到達した場合、勝ち抜け状態に設定
4. スコアが `win_point² - win_point` 以上かつ勝ち抜け直前の場合、リーチ状態に設定

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. スコアを `正答数 × (win_point - 誤答数)` で再計算
3. 誤答数が `lose_point` に到達した場合、失格状態に設定

#### スコア計算の詳細

```typescript
const newCorrect = playerState.correct + 1; // 正答時
const newScoreInCorrectCase =
  newCorrect * (game.win_point! - playerState.wrong);

const newWrong = playerState.wrong + 1; // 誤答時
const newScoreInWrongCase = playerState.correct * (game.win_point! - newWrong);
```

#### 勝利判定

```typescript
if (newScoreInCorrectCase >= game.win_point! ** 2) {
  // 勝ち抜け（N²に到達）
  state = "win";
}
```

#### 失格判定

```typescript
if (newWrong >= game.lose_point!) {
  // 失格（N回誤答）
  state = "lose";
}
```

### UI表示仕様

#### スコアボード表示

3行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時："LOSE"
  - 通常時：現在のスコア（例："12pt"）

- **2行目**：計算式表示（緑色ボタン、操作不可）
  - 形式：`{正答数}✕{N-誤答数}` （例："3✕4"）
  - 現在のスコア計算の内訳を表示

- **3行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色）
  - 右側：誤答ボタン（青色）

#### 色分け

- **メインボタン**：
  - 通常時：プレイ中を示す色（filled状態）
  - 勝ち抜け時：勝利状態の色
  - 失格時：失格状態の色

- **計算式ボタン**：常に緑色

- **操作ボタン**：正解は赤色、誤答は青色

### 設定可能パラメータ

#### 基準値N（win_point）

- 範囲：1-1000
- デフォルト：5
- 説明：スコア計算の基準となる値。N²で勝ち抜け

#### 失格誤答数（lose_point）

- 範囲：1-1000
- デフォルト：5
- 説明：この回数誤答すると失格

## 戦略的要素

### スコア上昇の特徴

1. **序盤の重要性**：誤答が少ない序盤ほどスコアの上昇率が高い
2. **誤答の重み**：誤答が増えるほど後の正答の価値が下がる
3. **リスク管理**：積極性と慎重さのバランスが重要

### 数学的特性

#### スコア上昇パターン（N=5の場合）

| 正解数 | 誤答0 | 誤答1 | 誤答2 | 誤答3 | 誤答4 |
| ------ | ----- | ----- | ----- | ----- | ----- |
| 1      | 5     | 4     | 3     | 2     | 1     |
| 2      | 10    | 8     | 6     | 4     | 2     |
| 3      | 15    | 12    | 9     | 6     | 3     |
| 4      | 20    | 16    | 12    | 8     | 4     |
| 5      | 25★   | 20    | 15    | 10    | 5     |

★勝ち抜け（25点）

### プレイヤーの戦略

1. **慎重型戦略**：誤答を避けて確実な問題のみ解答
2. **積極型戦略**：早期に正解を重ねて高スコアを狙う
3. **計算型戦略**：必要な正解数と許容誤答数を計算して解答

### 出題者の配慮

- 問題の難易度バランス調整
- 終盤での逆転要素の確保
- パラメータ設定による競技性の調整

## 他形式との比較

### nomx形式との違い

- **nomx**：固定回数での勝ち抜け・失格
- **nbyn**：変動するスコア計算による勝ち抜け

### ny形式との違い

- **ny**：単純な加減算
- **nbyn**：乗算を含む複雑な計算式

### squarex形式との違い

- **squarex**：奇数・偶数問での分離計算
- **nbyn**：すべての問題で統一計算

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- 乗算ベースの計算により、効率的なスコア更新を実現
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

## 使用例

### 設定例1：標準的な5by5設定

- 基準値N：5
- 失格誤答数：5
- 用途：一般的な競技クイズ

### 設定例2：短期戦3by3設定

- 基準値N：3
- 失格誤答数：3
- 用途：時間制限のある競技

### 設定例3：長期戦7by7設定

- 基準値N：7
- 失格誤答数：7
- 用途：実力重視の競技

## 実装上の課題とTODO

### 現在の制限事項

#### 1. リーチ状態の判定精度

**現状**：リーチ状態の判定ロジックが単純化されている可能性
**対策**：「もう1回正解で勝ち抜け」の条件をより正確に実装

#### 2. 数値オーバーフロー対策

**現状**：大きな数値での計算時の安全性が未検証
**対策**：極端な設定値での動作確認とエラーハンドリング

#### 3. 負のスコア処理

**現状**：誤答数がNを超えた場合の挙動が未定義
**対策**：スコアが負になる場合の表示・計算方法の明確化

### 将来的な改善提案

1. **設定バリデーション**：無効なパラメータ組み合わせの防止
2. **計算式表示の改善**：より直感的なスコア内訳表示
3. **統計情報の追加**：ゲーム終了後の詳細分析機能

この形式は、数学的な要素と戦略性を組み合わせた独特な競技性を持つルールとして設計されており、プレイヤーに計算能力と判断力の両方を求める挑戦的な形式です。
