# nomr形式（N○M休）仕様書

## 概要

nomr形式（N○M休、エヌマル・エヌヤス）は、競技クイズにおいて正答を重ねることで勝ち抜けを目指し、誤答時には一定期間の解答権剥奪というペナルティを課すルール形式です。誤答による失格はなく、すべてのプレイヤーが最後まで参加できる設計となっています。

## 基本ルール

### 勝利条件

- N回正解することで勝ち抜け
- デフォルト設定：7回正解で勝ち抜け

### ペナルティ条件

- 誤答するとM回「休み」状態になり、解答権を失う
- デフォルト設定：3問休み
- 何回誤答しても失格にはならない

### 休み状態の仕組み

- 誤答した時点から指定された問題数の間、解答権が剥奪される
- 休み回数が経過すると自動的に解答権が回復する
- 休み中も他のプレイヤーの解答を見ることは可能

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "nomr",
  win_point: number,    // 勝ち抜けに必要な正解数（デフォルト：7）
  lose_point: number,   // 休み回数（デフォルト：3）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは2つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | nomr形式での扱い       | 説明                               |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子       |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名                 |
| `initial_correct`    | `number` | 通常0（調整可能）      | スコア初期値として使用 | ゲーム開始時の正解数               |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 計算に影響なし         | ゲーム開始時の誤答数               |
| `base_correct_point` | `number` | 通常10                 | 使用されない           | 正解時の基本点数（nomrでは未使用） |
| `base_wrong_point`   | `number` | 通常-10                | 使用されない           | 誤答時の基本点数（nomrでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値            | 正解時           | 誤答時   | 休み明け時  | 説明                     |
| --------------- | --------- | ----------------- | ---------------- | -------- | ----------- | ------------------------ |
| `score`         | `number`  | `initial_correct` | 変更なし         | 変更なし | 変更なし    | nomr形式では使用されない |
| `correct`       | `number`  | `initial_correct` | `+1`             | 変更なし | 変更なし    | 現在の正解数             |
| `wrong`         | `number`  | `initial_wrong`   | 変更なし         | `+1`     | 変更なし    | 現在の誤答数             |
| `last_correct`  | `number`  | `-10`             | 問題番号         | 変更なし | 変更なし    | 最後に正解した問題番号   |
| `last_wrong`    | `number`  | `-10`             | 変更なし         | 問題番号 | 変更なし    | 最後に誤答した問題番号   |
| `is_incapacity` | `boolean` | `false`           | 変更なし         | `true`   | `false`     | 休み状態フラグ           |
| `state`         | `string`  | `"playing"`       | 勝利時`"win"`    | 変更なし | `"playing"` | プレイヤーの状態         |
| `reach_state`   | `string`  | `"normal"`        | リーチ時`"win"`  | 変更なし | `"normal"`  | リーチ状態               |
| `text`          | `string`  | 動的生成          | 動的更新         | 動的更新 | 動的更新    | 表示テキスト             |
| `order`         | `number`  | `0`               | 勝利時に順位設定 | 変更なし | 変更なし    | 勝ち抜け順位             |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. 正解数が`win_point`に到達した場合、勝ち抜け状態に設定
3. 正解数が`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. `is_incapacity`を`true`に設定（休み状態）
3. `last_wrong`に現在の問題番号を記録

#### 休み状態の回復判定

```typescript
// 休み状態から回復する条件
const isRecovered =
  game.lose_point! < gameLogList.length - playerState.last_wrong;
```

休み状態からの回復は、`logs`テーブルで`variant`が`correct`、`wrong`、`through`のいずれかであるログエントリの累積数に基づいて判定されます。最後に誤答した問題番号からの経過問題数が`lose_point`以上になると解答権が回復します。

**⚠️ 実装上の注意点**：
現在の実装では`gameLogList.length`を基準に`last_wrong`の値を更新していますが、この方法は他の操作（undo/redo等）によってログの順序や数が変わった場合にバグを生む可能性があります。より堅牢な実装が必要です。

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 休み中：残り休み回数（例："2休"）
  - 通常時：現在のポイント数（例："5pt"）

- **操作ボタン**
  - 正解ボタン：休み中はグレーアウトして無効化
  - 誤答ボタン：休み中はグレーアウトして無効化

- **色分け**
  - メインボタン：休み状態では青色、通常時は緑色
  - 勝ち抜け時：金色

#### レイアウト構成

2行構成でプレイヤー情報を表示：

- 上段：メイン状態表示
- 下段：正解・誤答操作ボタン

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：7
- 説明：この回数正解すると勝ち抜け

#### 休み回数（lose_point）

- 範囲：1-100
- デフォルト：3
- 説明：誤答時にこの問題数だけ休み状態になる

## 他形式との比較

### nomx形式との違い

- **nomx**：誤答で失格（永続的なペナルティ）
- **nomr**：誤答で休み（一時的なペナルティ）

### freezex形式との違い

- **freezex**：N回目の誤答でN問休み（累積的なペナルティ）
- **nomr**：毎回同じM問休み（固定的なペナルティ）

## 戦略的要素

### プレイヤーの戦略

1. **確実性重視**：休みになるリスクを避けて確実な問題のみ解答
2. **積極性重視**：休みを恐れず積極的に解答して先行を狙う
3. **タイミング重視**：他プレイヤーの状況を見て解答タイミングを調整

### 出題者の配慮

- 休み中のプレイヤーが多い場合の問題調整
- 終盤での逆転要素の考慮
- 休み回数の設定による難易度調整

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 最終正解問題番号の早さ
3. 現在のスコア
4. 正解数
5. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- 問題番号ベースでの休み計算により、効率的な状態判定を実現
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

## 使用例

### 設定例1：標準的な設定

- 勝ち抜けポイント：7
- 休み回数：3
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜けポイント：5
- 休み回数：2
- 用途：時間制限のある競技

### 設定例3：長期戦設定

- 勝ち抜けポイント：10
- 休み回数：4
- 用途：実力差を明確にしたい競技

この形式は、誤答に対して適度なペナルティを設けつつ、すべての参加者が最後まで競技に参加できるバランスの取れたルールとして設計されています。

## 実装上の課題とTODO

### バリデーション強化が必要な項目

#### 1. 休み中プレイヤーの誤答防止

**現状**：クライアント側でボタンを無効化しているが、サーバーサイドでのバリデーションが不十分
**対策**：休み状態のプレイヤーからの解答操作を受け付けないよう、バックエンドレベルでのバリデーションを追加する必要がある

#### 2. 勝ち抜け後プレイヤーの操作制限

**現状**：勝ち抜け後も操作が可能な状態
**対策**：勝ち抜け状態（`state: "win"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

#### 3. 休み回数計算の改善

**現状**：`gameLogList.length`ベースの計算でundo/redo時にバグの可能性
**対策**：より堅牢な問題番号管理システムの実装を検討する必要がある

### 将来的な改善提案

1. **複数回誤答時の挙動明確化**：休み中の誤答操作に対する仕様を明確に定義
2. **状態遷移の整合性チェック**：プレイヤー状態の不正な変更を防ぐ仕組み
3. **ログ管理の改善**：undo/redo操作に対してより安全な実装

## ユニットテスト観点

ユニットテストでは`getInitialPlayersStateForOnline`がnomr形式でプレイヤーごとの`initialScore`をそのままスコアと正解数に引き継ぎ、誤答数も同じ初期値で開始することを確認する。複数のプレイヤーに異なる初期正解数を設定し、全員が`is_incapacity: false`かつ`reach_state: "playing"`で初期化されること、並び替え時に初期正解数が高いプレイヤーが優先されること、休み状態を持たない初期ログで順位が安定して返却されることを検証する。
