# nomx-ad形式（連答つきN○M✕）仕様書

## 概要

nomx-ad形式（連答つきN○M✕、レントウツキ・エヌマル・エムバツ）は、基本的なnomx形式に連続正解によるアドバンテージ機能を追加した競技クイズ形式です。同じプレイヤーが連続で正解すると通常の2倍のポイントが獲得できるため、より戦略的でダイナミックな展開が期待できます。ABCクイズ大会の2nd Roundでよく採用されている形式です。

## 基本ルール

### 勝利条件

- 設定されたポイント数（win_point）に達すると勝ち抜け
- デフォルト設定：7ポイントで勝ち抜け

### 失格条件

- M回誤答することで失格
- デフォルト設定：3回誤答で失格
- 失格となったプレイヤーは以降の問題に参加できない

### アドバンテージシステム

- **通常状態**：正解時+1ポイント
- **アドバンテージ状態**：正解時+2ポイント
- **アドバンテージ獲得条件**：前問で正解したプレイヤーが次の問題も正解
- **アドバンテージ解除条件**：誤答、または他のプレイヤーが正解

### 連答オプション

#### streak_over3オプション

- **true**：3連答以上でもアドバンテージを継続
- **false**：2連答目以降はアドバンテージなし（デフォルト）

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "nomx-ad",
  win_point: number,    // 勝ち抜けに必要なポイント数（デフォルト：7）
  lose_point: number,   // 失格となる誤答数（デフォルト：3）
  win_through: number,  // 勝ち抜け人数（デフォルト：1）
  options: {
    streak_over3: boolean,  // 3連答以上でもアドバンテージ継続（デフォルト：true）
  }
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは2つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | nomx-ad形式での扱い    | 説明                         |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子 |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名           |
| `initial_correct`    | `number` | 通常0（調整可能）      | 正解数初期値として使用 | ゲーム開始時の正解数         |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 誤答数初期値として使用 | ゲーム開始時の誤答数         |
| `base_correct_point` | `number` | 通常1                  | 使用されない           | 正解時の基本点数（動的計算） |
| `base_wrong_point`   | `number` | 通常1                  | 誤答時の加算値         | 誤答時の基本点数             |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値            | 正解時              | 誤答時        | アドバンテージ時 | 勝ち抜け時     | 失格時        | 説明                       |
| -------------- | -------- | ----------------- | ------------------- | ------------- | ---------------- | -------------- | ------------- | -------------------------- |
| `score`        | `number` | `initial_correct` | `+1 or +2`          | 変更なし      | `+2`             | 変更なし       | 変更なし      | 現在のポイント数           |
| `correct`      | `number` | `initial_correct` | `+1`                | 変更なし      | `+1`             | 変更なし       | 変更なし      | 現在の正解数               |
| `wrong`        | `number` | `initial_wrong`   | 変更なし            | `+1`          | 変更なし         | 変更なし       | 変更なし      | 現在の誤答数               |
| `stage`        | `number` | `1`               | アドバンテージ時`2` | `1`にリセット | `2`を維持        | 変更なし       | `1`にリセット | アドバンテージ状態（2=AD） |
| `last_correct` | `number` | `-1`              | 問題番号            | 変更なし      | 問題番号         | 問題番号       | 変更なし      | 最後に正解した問題番号     |
| `last_wrong`   | `number` | `-1`              | 変更なし            | 問題番号      | 変更なし         | 変更なし       | 問題番号      | 最後に誤答した問題番号     |
| `state`        | `string` | `"playing"`       | `"playing"`         | `"playing"`   | `"playing"`      | `"win"`        | `"lose"`      | プレイヤーの状態           |
| `text`         | `string` | 動的生成          | 動的更新            | 動的更新      | 動的更新         | 順位表示       | `"LOSE"`      | 表示テキスト               |
| `order`        | `number` | `0`               | 変更なし            | 変更なし      | 変更なし         | 勝ち抜け順設定 | 変更なし      | 勝ち抜け順位               |

### 計算ロジック

#### 正答時の処理

```typescript
case "correct":
  const is_ad = playerState.stage === 2;  // アドバンテージ状態判定
  const newScore = playerState.score + (is_ad ? 2 : 1);  // スコア計算
  const next_ad =
    (!game.options.streak_over3 && playerState.stage === 1) ||
    game.options.streak_over3;  // 次問アドバンテージ判定

  last_correct_player = playerState.player_id;  // 最終正解者更新

  if (newScore >= game.win_point!) {
    // 勝ち抜け処理
    return {
      ...playerState,
      correct: playerState.correct + 1,
      score: newScore,
      last_correct: qn,
      state: "win",
      stage: next_ad ? 2 : 1,
    };
  } else {
    // 通常の正解処理
    return {
      ...playerState,
      correct: playerState.correct + 1,
      score: newScore,
      last_correct: qn,
      stage: next_ad ? 2 : 1,  // アドバンテージ状態設定
    };
  }
```

#### 誤答時の処理

```typescript
case "wrong":
  const newWrong = playerState.wrong + 1;

  // 最終正解者リセット
  if (last_correct_player === playerState.player_id) {
    last_correct_player = "";
  }

  return {
    ...playerState,
    wrong: newWrong,
    last_wrong: qn,
    state: newWrong >= game.lose_point! ? "lose" : "playing",
    stage: 1,  // 通常状態にリセット
  };
```

#### 他プレイヤー正解時の処理

```typescript
// 他のプレイヤーが正解した場合、アドバンテージ状態を解除
if (log.variant === "correct" && playerState.stage === 2) {
  return {
    ...playerState,
    stage: 1, // アドバンテージ解除
  };
}
```

### アドバンテージシステムの詳細

#### stage値の意味

- **stage: 1**：通常状態（正解時+1ポイント）
- **stage: 2**：アドバンテージ状態（正解時+2ポイント）

#### アドバンテージ獲得の条件分岐

```typescript
const next_ad =
  (!game.options.streak_over3 && playerState.stage === 1) ||
  game.options.streak_over3;
```

- **streak_over3が false**：初回正解時のみアドバンテージ獲得
- **streak_over3が true**：連続正解している限りアドバンテージ継続

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時：「LOSE」表示
  - 通常時：現在のポイント数（例：「5pt」）

- **アドバンテージ表示**
  - stage=2の場合、視覚的にアドバンテージ状態を表示
  - 背景色やボーダーで区別

- **操作ボタン**
  - 正解ボタン：失格後はグレーアウトして無効化
  - 誤答ボタン：失格後はグレーアウトして無効化

- **色分け**
  - リーチ時：プレイヤー名のボーダーカラーが赤色
  - アドバンテージ時：特別な色表示
  - 勝ち抜け時：金色
  - 失格時：グレーアウト

#### レイアウト構成

3行構成でプレイヤー情報を表示：

- 上段：メイン状態表示
- 中段：アドバンテージ状態表示
- 下段：正解・誤答操作ボタン

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：7
- 説明：このポイント数に達すると勝ち抜け

#### 失格ポイント（lose_point）

- 範囲：1-100
- デフォルト：3
- 説明：この回数誤答すると失格

#### 連答オプション（streak_over3）

- **true**：3連答以上でもアドバンテージを継続
- **false**：2連答目以降はアドバンテージなし
- デフォルト：true
- 説明：連続正解時のアドバンテージ継続設定

## 他形式との比較

### nomx形式との違い

- **nomx**：正解数のみカウント、固定+1ポイント
- **nomx-ad**：ポイント制、連答により+2ポイント獲得可能

### ny形式との違い

- **nomx-ad**：誤答でのマイナスなし、失格制
- **ny**：誤答で-1ポイント、失格なし

### 戦術的な違い

- 連答狙いによる積極的な解答が有効
- アドバンテージ状態の相手を止める戦術の重要性

## 戦略的要素

### プレイヤーの戦略

1. **連答狙い**：アドバンテージ獲得のため積極的に連続正解を狙う
2. **相手潰し**：アドバンテージ状態の相手を阻止する解答
3. **リスク管理**：失格リスクとアドバンテージ獲得のバランス

### 特殊な戦術

1. **ストップ戦術**：相手のアドバンテージを意図的に阻止
2. **温存戦術**：確実な問題でアドバンテージを狙う
3. **カウンター戦術**：相手のアドバンテージ失効直後を狙う

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のポイント数
3. 最終正解問題番号の早さ
4. 正解数
5. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- 連答判定のためのプレイヤー状態追跡
- アドバンテージ状態の効率的な管理
- UI更新は状態変更時のみ実行

## 使用例

### 設定例1：ABC 2nd Round標準設定

- 勝ち抜けポイント：7
- 失格ポイント：3
- streak_over3：true
- 用途：ABC大会2nd Round

### 設定例2：短期戦設定

- 勝ち抜けポイント：5
- 失格ポイント：2
- streak_over3：false
- 用途：時間制限のある競技

### 設定例3：連答重視設定

- 勝ち抜けポイント：10
- 失格ポイント：4
- streak_over3：true
- 用途：連答技術を重視したい競技

この形式は、基本的なnomx形式に戦略性とダイナミズムを加えた、より高度な競技クイズ形式として設計されています。連答によるアドバンテージシステムにより、プレイヤー間の駆け引きがより複雑になり、観戦者にとっても楽しめる形式となっています。

## ABCクイズ大会での使用実績

### 採用背景

ABCクイズ大会の2nd Roundでは、1st Roundを通過したプレイヤーが更なる実力を競う場として、この連答つきnomx形式が採用されています。

### 特徴的な使用法

- **連答の重要性**：2連答以上でアドバンテージを獲得できるため、知識の幅と確実性が同時に求められる
- **戦術的多様性**：相手のアドバンテージを阻止する戦術や、確実な連答を狙う戦術など多彩な戦略が生まれる
- **観戦の面白さ**：アドバンテージ状態の表示により、観戦者にとっても状況が分かりやすい

## 実装上の課題とTODO

### バリデーション強化が必要な項目

#### 1. アドバンテージ状態の整合性チェック

**現状**：アドバンテージ状態の設定・解除が複数の条件で行われるため、整合性の確認が困難
**対策**：アドバンテージ状態遷移の単体テストを充実させ、edge caseでの動作を保証する

#### 2. 最終正解者の管理

**現状**：グローバル変数での管理により、並行処理時の競合状態の可能性
**対策**：より安全な状態管理機構の導入を検討

### 将来的な改善提案

1. **アドバンテージ表示の強化**：連答数やアドバンテージ履歴の表示
2. **カスタム連答ルール**：3連答、4連答等でのさらなるボーナス設定
3. **統計情報の充実**：連答成功率、アドバンテージ活用率等の詳細統計
