# nomx形式（N○M✕）仕様書

## 概要

nomx形式（N○M✕、エヌマル・エムバツ）は、競技クイズにおいて最も基本的な形式の一つです。N回正解で勝ち抜け、M回誤答で失格という分かりやすいルールが特徴で、代表的なものに「ナナマルサンバツ（7○3✕）」があります。誤答に対して厳格なペナルティを設ける、緊張感のある競技形式です。

## 基本ルール

### 勝利条件

- N回正解することで勝ち抜け
- デフォルト設定：7回正解で勝ち抜け

### 失格条件

- M回誤答することで失格
- デフォルト設定：3回誤答で失格
- 失格となったプレイヤーは以降の問題に参加できない

### ゲーム終了

- 設定された人数が勝ち抜けるか、全問題が終了した時点でゲーム終了

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "nomx",
  win_point: number,    // 勝ち抜けに必要な正解数（デフォルト：7）
  lose_point: number,   // 失格となる誤答数（デフォルト：3）
  win_through: number,  // 勝ち抜け人数（デフォルト：1）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは2つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型 | 初期値 | 説明 |
|------------|---|--------|------|
| `id` | `string` | 自動生成UUID | プレイヤーの一意識別子 |
| `name` | `string` | ユーザー入力 | プレイヤー名 |
| `text` | `string` | ユーザー入力 | プレイヤーの説明文・備考 |
| `belong` | `string` | ユーザー入力 | 所属団体・学校など |
| `tags` | `string[]` | `[]` | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名 | 型 | 初期値 | nomx形式での扱い | 説明 |
|------------|---|--------|-----------------|------|
| `id` | `string` | マスターデータより継承 | プレイヤー識別に使用 | ゲーム内でのプレイヤー識別子 |
| `name` | `string` | マスターデータより継承 | 表示に使用 | ゲーム内での表示名 |
| `initial_correct` | `number` | 通常0（調整可能） | 正解数初期値として使用 | ゲーム開始時の正解数 |
| `initial_wrong` | `number` | 通常0（調整可能） | 誤答数初期値として使用 | ゲーム開始時の誤答数 |
| `base_correct_point` | `number` | 通常1 | 正解時の加算値 | 正解時の基本点数 |
| `base_wrong_point` | `number` | 通常1 | 誤答時の加算値 | 誤答時の基本点数 |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名 | 型 | 初期値 | 正解時 | 誤答時 | 勝ち抜け時 | 失格時 | 説明 |
|------------|---|--------|--------|--------|-----------|-------|------|
| `score` | `number` | `initial_correct` | 変更なし | 変更なし | 変更なし | 変更なし | nomx形式では使用されない |
| `correct` | `number` | `initial_correct` | `+1` | 変更なし | 変更なし | 変更なし | 現在の正解数 |
| `wrong` | `number` | `initial_wrong` | 変更なし | `+1` | 変更なし | 変更なし | 現在の誤答数 |
| `last_correct` | `number` | `-1` | 問題番号 | 変更なし | 問題番号 | 変更なし | 最後に正解した問題番号 |
| `last_wrong` | `number` | `-1` | 変更なし | 問題番号 | 変更なし | 問題番号 | 最後に誤答した問題番号 |
| `state` | `string` | `"playing"` | `"playing"` | `"playing"` | `"win"` | `"lose"` | プレイヤーの状態 |
| `reach_state` | `string` | `"normal"` | リーチ時設定 | リーチ時設定 | `"normal"` | `"normal"` | リーチ状態 |
| `text` | `string` | 動的生成 | 動的更新 | 動的更新 | 順位表示 | `"LOSE"` | 表示テキスト |
| `order` | `number` | `0` | 変更なし | 変更なし | 勝ち抜け順設定 | 変更なし | 勝ち抜け順位 |

### 計算ロジック

#### 正答時の処理

```typescript
case "correct":
  const newCorrect = playerState.correct + 1;
  if (newCorrect >= game.win_point!) {
    // 勝ち抜け処理
    return {
      ...playerState,
      correct: newCorrect,
      last_correct: qn,
      state: "win",
    };
  } else if (newCorrect + 1 === game.win_point!) {
    // 勝ち抜けリーチ状態設定
    return {
      ...playerState,
      correct: newCorrect,
      last_correct: qn,
      reach_state: "win",
    };
  }
```

1. プレイヤーの正解数を1増加
2. 正解数が`win_point`に到達した場合、勝ち抜け状態に設定
3. 正解数が`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定

#### 誤答時の処理

```typescript
case "wrong":
  const newWrong = playerState.wrong + 1;
  if (newWrong >= game.lose_point!) {
    // 失格処理
    return {
      ...playerState,
      wrong: newWrong,
      last_wrong: qn,
      state: "lose",
    };
  } else if (
    newWrong + 1 === game.lose_point! &&
    playerState.correct + 1 !== game.win_point!
  ) {
    // 失格リーチ状態設定（勝ち抜けリーチでない場合のみ）
    return {
      ...playerState,
      wrong: newWrong,
      last_wrong: qn,
      reach_state: "lose",
    };
  }
```

1. プレイヤーの誤答数を1増加
2. 誤答数が`lose_point`に到達した場合、失格状態に設定
3. 誤答数が`lose_point - 1`に到達し、かつ勝ち抜けリーチでない場合、失格リーチ状態に設定

**⚠️ リーチ状態の優先順位**：
勝ち抜けリーチと失格リーチが同時に発生する可能性がある場合、勝ち抜けリーチが優先されます。

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時：「LOSE」表示
  - 通常時：「正解数 - 誤答数」形式（例：「3 - 1」）

- **操作ボタン**
  - 正解ボタン：失格後はグレーアウトして無効化
  - 誤答ボタン：失格後はグレーアウトして無効化

- **色分け**
  - リーチ時：プレイヤー名のボーダーカラーが赤色
  - 勝ち抜け時：金色
  - 失格時：グレーアウト

#### レイアウト構成

2行構成でプレイヤー情報を表示：

- 上段：メイン状態表示
- 下段：正解・誤答操作ボタン

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：7
- 説明：この回数正解すると勝ち抜け

#### 失格ポイント（lose_point）

- 範囲：1-100
- デフォルト：3
- 説明：この回数誤答すると失格

#### 勝ち抜け人数（win_through）

- 範囲：1-プレイヤー数
- デフォルト：1
- 説明：この人数が勝ち抜けた時点でゲーム終了

## 他形式との比較

### nomr形式との違い

- **nomx**：誤答で失格（永続的なペナルティ）
- **nomr**：誤答で休み（一時的なペナルティ）

### ny形式との違い

- **nomx**：正解のみカウント、誤答は失格要因
- **ny**：正解で+1、誤答で-1の加減式

### nupdown形式との違い

- **nomx**：正解数と誤答数を独立して管理
- **nupdown**：誤答時に正解数をリセット

## 戦略的要素

### プレイヤーの戦略

1. **確実性重視**：失格リスクを避けて確実な問題のみ解答
2. **積極性重視**：リスクを取って早期勝ち抜けを狙う
3. **相手観察**：他プレイヤーの状況を見て解答タイミングを調整

### 出題者の配慮

- 難易度バランスの調整
- 失格者が多くなった場合の問題選択
- リーチ者への適切な問題提供

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 最終正解問題番号の早さ
3. 現在の正解数
4. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- シンプルなカウンター計算により高速な状態判定を実現
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

## 使用例

### 設定例1：標準的なナナマルサンバツ

- 勝ち抜けポイント：7
- 失格ポイント：3
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜けポイント：5
- 失格ポイント：2
- 用途：時間制限のある競技

### 設定例3：緊張感重視設定

- 勝ち抜けポイント：10
- 失格ポイント：2
- 用途：誤答に対して厳しいペナルティを課したい競技

この形式は、明確でシンプルなルールでありながら、プレイヤーに高い緊張感を提供する競技クイズの王道形式として広く愛用されています。

## 実装上の課題とTODO

### バリデーション強化が必要な項目

#### 1. 失格後プレイヤーの操作制限

**現状**：クライアント側でボタンを無効化しているが、サーバーサイドでのバリデーションが不十分
**対策**：失格状態（`state: "lose"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

#### 2. 勝ち抜け後プレイヤーの操作制限

**現状**：勝ち抜け後も操作が可能な状態
**対策**：勝ち抜け状態（`state: "win"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

### 将来的な改善提案

1. **多段階リーチ表示**：2問前、1問前などの段階的なリーチ表示
2. **統計情報の充実**：勝率、平均正解数等の詳細統計
3. **カスタムルール対応**：特殊な勝利条件の設定機能