# Nupdown形式仕様書

## 概要

Nupdown形式（エヌアップダウン）は、競技クイズにおいて誤答に対して極めて厳しいペナルティを課すルール形式です。正解することでスコアが1ずつ上昇し、目標ポイントに達することで勝ち抜けを目指しますが、一度でも誤答するとスコアが0にリセットされるという特徴を持ちます。このため、一回の誤答が結果を大きく左右する、誤答に非常に厳しい形式として設計されています。

## 基本ルール

### 勝利条件

- スコアがN点に達することで勝ち抜け
- デフォルト設定：5点で勝ち抜け

### 失格条件

- M回誤答すると失格
- デフォルト設定：2回誤答で失格

### スコア計算システム

- **正解時**：スコア+1
- **誤答時**：スコアが0にリセット（何点あっても必ず0になる）

### リーチ状態

- **勝ち抜けリーチ**：スコアが目標点数-1に達した状態
- **失格リーチ**：誤答回数が失格条件-1に達した状態

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "nupdown",
  win_point: number,    // 勝ち抜けに必要なスコア（デフォルト：5）
  lose_point: number,   // 失格となる誤答回数（デフォルト：2）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | nupdown形式での扱い    | 説明                                   |
| -------------------- | -------- | ---------------------- | ---------------------- | -------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子           |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名                     |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の正解数                   |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の誤答数                   |
| `base_correct_point` | `number` | 通常1                  | 使用されない           | 正解時の基本点数（nupdownでは未使用）  |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない           | 誤答時の基本点数（nupdownでは未使用）  |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値      | 正解時           | 誤答時   | 説明                     |
| --------------- | --------- | ----------- | ---------------- | -------- | ------------------------ |
| `score`         | `number`  | `0`         | `+1`             | `0`にリセット | 現在のスコア             |
| `correct`       | `number`  | `0`         | `+1`             | 変更なし | 現在の正解数             |
| `wrong`         | `number`  | `0`         | 変更なし         | `+1`     | 現在の誤答数             |
| `last_correct`  | `number`  | `-1`        | 問題番号         | 変更なし | 最後に正解した問題番号   |
| `last_wrong`    | `number`  | `-1`        | 変更なし         | 問題番号 | 最後に誤答した問題番号   |
| `state`         | `string`  | `"playing"` | 勝利時`"win"`    | 失格時`"lose"` | プレイヤーの状態         |
| `reach_state`   | `string`  | `"normal"`  | リーチ時`"win"`  | リーチ時`"lose"` | リーチ状態               |
| `text`          | `string`  | 動的生成    | 動的更新         | 動的更新 | 表示テキスト             |
| `order`         | `number`  | `0`         | 勝利時に順位設定 | 変更なし | 勝ち抜け順位             |

### 計算ロジック

#### 正答時の処理

1. プレイヤーのスコアを1増加
2. プレイヤーの正解数を1増加
3. スコアが`win_point`に到達した場合、勝ち抜け状態に設定
4. スコアが`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定

```typescript
const newScore = playerState.score + 1;
const newCorrect = playerState.correct + 1;
if (newScore >= game.win_point!) {
  // 勝ち抜け
  state = "win";
} else if (newScore + 1 === game.win_point!) {
  // 勝ち抜けリーチ
  reach_state = "win";
}
```

#### 誤答時の処理

1. プレイヤーのスコアを0にリセット
2. プレイヤーの誤答数を1増加
3. 誤答数が`lose_point`に到達した場合、失格状態に設定
4. 誤答数が`lose_point - 1`に到達した場合かつ勝ち抜けリーチでない場合、失格リーチ状態に設定

```typescript
const newWrong = playerState.wrong + 1;
if (newWrong >= game.lose_point!) {
  // 失格
  state = "lose";
  score = 0;
} else if (
  newWrong + 1 === game.lose_point! &&
  playerState.score + 1 !== game.win_point!
) {
  // 失格リーチ（勝ち抜けリーチでない場合のみ）
  reach_state = "lose";
  score = 0;
} else {
  // 通常の誤答（スコアリセットのみ）
  score = 0;
}
```

### UI表示仕様

#### スコアボード表示

2行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時："LOSE"
  - 通常時：現在のスコア（例："3pt"）

- **2行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）

#### 色分けとリーチ表示

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色
  - 失格時：失格状態の色
  - **勝ち抜けリーチ時**：赤色のボーダー表示で強調

- **操作ボタン**：正解は赤色、誤答は青色

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← スコア・状態表示
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：5
- 説明：この点数に達すると勝ち抜け

#### 失格誤答数（lose_point）

- 範囲：1-1000
- デフォルト：2
- 説明：この回数誤答すると失格

## 戦略的要素

### 誤答の重大性

Nupdown形式の最大の特徴は、誤答によるスコアリセットです：

1. **一度の誤答の影響**：4点まで積み上げても、1回の誤答で0点に戻る
2. **累積効果**：誤答により、それまでの正解がすべて無駄になる
3. **心理的プレッシャー**：高スコア時ほど誤答のリスクが重大になる

### 数学的特性

#### スコア進行パターン（5点勝ち抜け、2誤答失格の場合）

| 状況 | 必要正解数 | リスク |
|------|------------|--------|
| 0点時 | 5回連続正解 | 低 |
| 1点時 | 4回連続正解 | 中 |
| 2点時 | 3回連続正解 | 中 |
| 3点時 | 2回連続正解 | 高 |
| 4点時 | 1回正解（リーチ） | 最高 |

### プレイヤーの戦略

1. **超慎重型戦略**：確実な問題のみ解答し、リスクを最小化
2. **計算型戦略**：自分の正答率を考慮した確率的判断
3. **積極型戦略**：早期決着を狙い、リスクを取って解答
4. **状況判断型戦略**：他プレイヤーの状況を見て戦略を変更

### 心理的要素

- **プレッシャー管理**：高スコア時の心理的負担
- **リスク許容度**：個人の性格による戦略の違い
- **集中力維持**：長時間の緊張状態への対応

## 他形式との比較

### nomx形式との違い

- **nomx**：誤答により失格（永続的ペナルティ）
- **nupdown**：誤答によりスコアリセット（回復可能だが厳しい）

### ny形式との違い

- **ny**：誤答で-1点（緩やかなペナルティ）
- **nupdown**：誤答で0点リセット（厳格なペナルティ）

### nomr形式との違い

- **nomr**：誤答で休み（一時的ペナルティ）
- **nupdown**：誤答でリセット（即座の重大なペナルティ）

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- スコアリセットによる単純な計算処理
- UI更新は状態変更時のみ実行され、不要な再描画を抑制
- リーチ状態の効率的な判定

## 使用例

### 設定例1：標準的な5updown設定

- 勝ち抜けポイント：5
- 失格誤答数：2
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦3updown設定

- 勝ち抜けポイント：3
- 失格誤答数：2
- 用途：時間制限のある競技

### 設定例3：超厳格設定

- 勝ち抜けポイント：7
- 失格誤答数：1
- 用途：極めて高い精度を要求する競技

### 設定例4：やや寛容設定

- 勝ち抜けポイント：5
- 失格誤答数：3
- 用途：学習目的の競技

## 競技性の特徴

### 適用場面

1. **高精度を重視する競技**：誤答を極力避ける能力を測定
2. **プレッシャー耐性の測定**：高スコア時の判断能力を評価
3. **リスク管理能力の評価**：確率的思考と決断力を重視

### 出題者の配慮

- **問題難易度の調整**：極端に難しい問題は避ける
- **ボーダーライン問題の配慮**：解答者が判断に迷う問題の扱い
- **時間設定**：十分な思考時間の確保

## 実装上の課題とTODO

### 現在の制限事項

#### 1. リーチ状態の視覚的表現

**現状**：勝ち抜けリーチの赤ボーダー表示
**検討点**：失格リーチの表現方法の統一

#### 2. スコアリセット時のフィードバック

**現状**：単純なスコア表示更新
**改善案**：リセット時のアニメーション効果

#### 3. 戦略ヒントの提供

**現状**：基本的なスコア表示のみ
**改善案**：勝率計算やリスク評価の表示

### 将来的な改善提案

1. **統計情報の充実**：正答率や平均スコアの表示
2. **リプレイ機能**：重要な場面の振り返り
3. **難易度調整機能**：動的な問題選択システム
4. **心理的サポート**：プレッシャー軽減のための表示オプション

この形式は、競技クイズにおいて「一度の誤答も許されない緊張感」を作り出す独特なルールとして設計されており、プレイヤーの慎重さと正確性を最大限に引き出す挑戦的な形式です。