# ny形式（NewYork）仕様書

## 概要

ny形式（NewYork、ニューヨーク）は、競技クイズにおいて正答時に得点を獲得し、誤答時に得点を失うことで目標得点を目指すルール形式です。スコアが上下に変動するため、逆転要素が強く、最後まで結果が読めない緊迫感のある形式となっています。

## 基本ルール

### 勝利条件

- N点に到達することで勝ち抜け
- デフォルト設定：10点で勝ち抜け

### 敗退条件

- M回誤答することで失格
- デフォルト設定：敗退条件なし（制限なしで続行）

### スコア変動

- 正答：+1点
- 誤答：-1点
- 各プレイヤーは0点からスタート

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "ny",
  win_point: number,    // 勝ち抜けに必要な得点（デフォルト：10）
  lose_point?: number,  // 敗退に必要な誤答数（オプション）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは2つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | ny形式での扱い         | 説明                             |
| -------------------- | -------- | ---------------------- | ---------------------- | -------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子     |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名               |
| `initial_correct`    | `number` | 通常0（調整可能）      | 正解数初期値として使用 | ゲーム開始時の正解数             |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 誤答数初期値として使用 | ゲーム開始時の誤答数             |
| `base_correct_point` | `number` | 通常10                 | 使用されない           | 正解時の基本点数（nyでは未使用） |
| `base_wrong_point`   | `number` | 通常-10                | 使用されない           | 誤答時の基本点数（nyでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値            | 正解時                             | 誤答時         | 説明                     |
| -------------- | -------- | ----------------- | ---------------------------------- | -------------- | ------------------------ |
| `score`        | `number` | `0`               | `+1`                               | `-1`           | 現在の得点（メイン表示） |
| `correct`      | `number` | `initial_correct` | `+1`                               | 変更なし       | 現在の正解数             |
| `wrong`        | `number` | `initial_wrong`   | 変更なし                           | `+1`           | 現在の誤答数             |
| `last_correct` | `number` | `-10`             | 問題番号                           | 変更なし       | 最後に正解した問題番号   |
| `last_wrong`   | `number` | `-10`             | 変更なし                           | 問題番号       | 最後に誤答した問題番号   |
| `state`        | `string` | `"playing"`       | 勝利時`"win"`<br/>敗退時`"lose"`   | 敗退時`"lose"` | プレイヤーの状態         |
| `reach_state`  | `string` | `"playing"`       | リーチ時`"win"`<br/>危険時`"lose"` | 危険時`"lose"` | リーチ状態               |
| `text`         | `string` | 動的生成          | 動的更新                           | 動的更新       | 表示テキスト             |
| `order`        | `number` | `0`               | 勝利時に順位設定                   | 変更なし       | 勝ち抜け順位             |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. プレイヤーのスコアを1増加
3. スコアが`win_point`に到達した場合、勝ち抜け状態に設定
4. スコアが`win_point - 1`に到達した場合、勝ち抜けリーチ状態に設定

```typescript
case "correct":
  const newCorrect = playerState.correct + 1;
  const newScore = playerState.score + 1;
  if (newScore >= game.win_point!) {
    return {
      ...playerState,
      score: newScore,
      correct: newCorrect,
      last_correct: qn,
      state: "win",
    };
  } else if (newScore === game.win_point!) {
    return {
      ...playerState,
      score: newScore,
      correct: newCorrect,
      last_correct: qn,
      reach_state: "win",
    };
  }
```

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. プレイヤーのスコアを1減少
3. `lose_point`が設定されている場合：
   - 誤答数が`lose_point`に到達すると敗退状態に設定
   - 誤答数が`lose_point - 1`に到達すると敗退リーチ状態に設定

```typescript
case "wrong":
  const newWrong = playerState.wrong + 1;
  if (newWrong >= game.lose_point!) {
    return {
      ...playerState,
      score: playerState.score - 1,
      wrong: newWrong,
      last_wrong: qn,
      state: "lose",
    };
  } else if (
    newWrong + 1 === game.lose_point! &&
    playerState.correct + 1 !== game.win_point!
  ) {
    return {
      ...playerState,
      score: playerState.score - 1,
      wrong: newWrong,
      last_wrong: qn,
      reach_state: "lose",
    };
  }
```

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 敗退時："LOSE"
  - 通常時：現在の得点（例："7pt"）

- **操作ボタン**
  - 正解ボタン：勝ち抜け・敗退後はグレーアウトして無効化
  - 誤答ボタン：勝ち抜け・敗退後はグレーアウトして無効化

- **色分け**
  - メインボタン：勝ち抜けリーチ時は金色、敗退リーチ時は赤色、通常時は緑色
  - 勝ち抜け時：金色
  - 敗退時：赤色

#### レイアウト構成

2行構成でプレイヤー情報を表示：

- 上段：メイン状態表示（得点またはステータス）
- 下段：正解・誤答操作ボタン

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：10
- 説明：この得点に到達すると勝ち抜け

#### 敗退ポイント（lose_point）

- 範囲：1-100（オプション）
- デフォルト：未設定（敗退条件なし）
- 説明：この回数誤答すると敗退

## 他形式との比較

### normal形式との違い

- **normal**：得点のみ管理、勝敗判定なし
- **ny**：得点管理＋勝敗判定あり、スコアが上下動

### nomx形式との違い

- **nomx**：正解・誤答回数による勝敗判定、スコア概念なし
- **ny**：得点による勝敗判定、正解で+1・誤答で-1のスコア変動

### swedish10形式との違い

- **swedish10**：正解で+1、誤答で-1、10点で勝ち抜け（同一コンセプト）
- **ny**：基本的にswedish10と同じロジックだが、敗退条件の実装が異なる

## 戦略的要素

### プレイヤーの戦略

1. **積極戦略**：リスクを取って早期の勝ち抜けを狙う
2. **堅実戦略**：誤答を避けて着実に得点を重ねる
3. **逆転戦略**：後半での一気に得点獲得を狙う
4. **守備戦略**：リーチ後は確実な問題のみ解答

### スコア変動の特徴

- **逆転要素が強い**：1回の誤答で得点が下がるため、終盤まで結果が読めない
- **心理的プレッシャー**：得点が下がることで精神的な負担が大きい
- **連続正解の重要性**：得点を積み重ねるには連続正解が重要

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- スコア計算は単純な加減算のため、計算負荷が軽い
- 状態変更時のみUI更新を行い、不要な再描画を抑制

## 使用例

### 設定例1：標準的な設定

- 勝ち抜けポイント：10
- 敗退ポイント：未設定
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜けポイント：7
- 敗退ポイント：未設定
- 用途：時間制限のある競技

### 設定例3：敗退ありの設定

- 勝ち抜けポイント：10
- 敗退ポイント：5
- 用途：誤答にペナルティを設けたい競技

この形式は、得点の上下動による緊張感と逆転要素を重視した、エキサイティングなクイズ形式として設計されています。

## 実装上の課題とTODO

### バリデーション強化が必要な項目

#### 1. 勝ち抜け後プレイヤーの操作制限

**現状**：勝ち抜け後も操作が可能な状態
**対策**：勝ち抜け状態（`state: "win"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

#### 2. 敗退後プレイヤーの操作制限

**現状**：敗退後も操作が可能な状態
**対策**：敗退状態（`state: "lose"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

#### 3. スコアのマイナス値制限

**現状**：スコアがマイナス値になる可能性
**対策**：スコアの下限値を設定するかどうか仕様を明確化する必要がある

### テスト不足の解決

#### 1. 単体テストの追加

**現状**：ny形式専用のテストファイル（`ny.test.ts`）が存在しない
**対策**：正解・誤答時のスコア変動、勝利・敗退条件、リーチ状態の判定をテストする必要がある

#### 2. 統合テストの追加

**現状**：`core.test.ts`にny形式のテストケースが含まれていない
**対策**：他の形式と同様にny形式の統合テストを追加する必要がある

### 将来的な改善提案

1. **スコア表示の改善**：マイナススコア時の表示方法を検討
2. **敗退条件の柔軟化**：スコアによる敗退条件の追加も検討
3. **状態遷移の整合性チェック**：プレイヤー状態の不正な変更を防ぐ仕組み
