# SquareX形式仕様書

## 概要

SquareX形式（スクエアエックス）は、競技クイズにおいて奇数問目と偶数問目の正解数をそれぞれ独立して管理し、その積（掛け算）によってスコアを算出する独特なルール形式です。奇数問目での正解と偶数問目での正解を両方バランスよく取る必要があり、片方だけでは高スコアが得られないため、戦略的な解答タイミングの判断が重要となる二次元的な競技形式です。

## 基本ルール

### 勝利条件

- 奇数問目正解数 × 偶数問目正解数 がX以上になると勝ち抜け
- デフォルト設定：16以上で勝ち抜け

### 失格条件

- 失格条件は存在しない

### スコア計算システム

- **奇数問目での正解**：奇数スコア+1
- **偶数問目での正解**：偶数スコア+1
- **最終スコア**：奇数スコア × 偶数スコア
- **誤答の扱い**：スコアに影響なし（誤答数のカウントのみ）

### 問題番号の分類

- **奇数問目**：0番目、2番目、4番目、... の問題
- **偶数問目**：1番目、3番目、5番目、... の問題

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "squarex",
  win_point: number,    // 勝ち抜けに必要なスコア（デフォルト：16）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | squarex形式での扱い  | 説明                                  |
| -------------------- | -------- | ---------------------- | -------------------- | ------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用 | ゲーム内でのプレイヤー識別子          |
| `name`               | `string` | マスターデータより継承 | 表示に使用           | ゲーム内での表示名                    |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない         | ゲーム開始時の正解数                  |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない         | ゲーム開始時の誤答数                  |
| `base_correct_point` | `number` | 通常1                  | 使用されない         | 正解時の基本点数（squarexでは未使用） |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない         | 誤答時の基本点数（squarexでは未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名   | 型       | 初期値      | 奇数問正解時     | 偶数問正解時     | 誤答時   | 説明                       |
| -------------- | -------- | ----------- | ---------------- | ---------------- | -------- | -------------------------- |
| `score`        | `number` | `0`         | 計算により更新   | 計算により更新   | 変更なし | 奇数×偶数スコアの積        |
| `correct`      | `number` | `0`         | `+1`             | `+1`             | 変更なし | 現在の正解数               |
| `wrong`        | `number` | `0`         | 変更なし         | 変更なし         | `+1`     | 現在の誤答数               |
| `odd_score`    | `number` | `0`         | `+1`             | 変更なし         | 変更なし | 奇数問目での正解数         |
| `even_score`   | `number` | `0`         | 変更なし         | `+1`             | 変更なし | 偶数問目での正解数         |
| `last_correct` | `number` | `-1`        | 問題番号         | 問題番号         | 変更なし | 最後に正解した問題番号     |
| `last_wrong`   | `number` | `-1`        | 変更なし         | 変更なし         | 問題番号 | 最後に誤答した問題番号     |
| `state`        | `string` | `"playing"` | 勝利時`"win"`    | 勝利時`"win"`    | 変更なし | プレイヤーの状態           |
| `reach_state`  | `string` | `"playing"` | 変更なし         | 変更なし         | 変更なし | リーチ状態（使用されない） |
| `text`         | `string` | 動的生成    | 動的更新         | 動的更新         | 動的更新 | 表示テキスト               |
| `order`        | `number` | `0`         | 勝利時に順位設定 | 勝利時に順位設定 | 変更なし | 勝ち抜け順位               |

### 計算ロジック

#### 正答時の処理

1. 問題番号が偶数（0, 2, 4, ...）の場合、奇数スコアを1増加
2. 問題番号が奇数（1, 3, 5, ...）の場合、偶数スコアを1増加
3. 新しいスコアを奇数スコア × 偶数スコアで計算
4. スコアが`win_point`以上になった場合、勝ち抜け状態に設定

```typescript
let newOddScore = playerState.odd_score;
let newEvenScore = playerState.even_score;

if (qn % 2 === 0) {
  // 奇数問目（0番目、2番目、...）
  newOddScore++;
} else {
  // 偶数問目（1番目、3番目、...）
  newEvenScore++;
}

const newScore = newOddScore * newEvenScore;
if (newScore >= game.win_point!) {
  // 勝ち抜け
  state = "win";
}
```

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. スコアには影響しない

#### 具体的な計算例

**16点勝ち抜けの場合**

| 問題     | 結果 | 奇数スコア | 偶数スコア | 最終スコア    |
| -------- | ---- | ---------- | ---------- | ------------- |
| 1問目(0) | 正解 | 1          | 0          | 0             |
| 2問目(1) | 正解 | 1          | 1          | 1             |
| 3問目(2) | 正解 | 2          | 1          | 2             |
| 4問目(3) | 正解 | 2          | 2          | 4             |
| 5問目(4) | 正解 | 3          | 2          | 6             |
| 6問目(5) | 正解 | 3          | 3          | 9             |
| 7問目(6) | 正解 | 4          | 3          | 12            |
| 8問目(7) | 正解 | 4          | 4          | 16 (勝ち抜け) |

### UI表示仕様

#### スコアボード表示

3行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 通常時：現在のスコア（例："12pt"）

- **2行目**：計算式表示（緑色ボタン、操作不可）
  - 形式：`{奇数スコア}✕{偶数スコア}` （例："4✕3"）
  - 現在のスコア計算の内訳を表示

- **3行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）

#### 色分け

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色

- **計算式ボタン**：常に緑色、filled状態

- **操作ボタン**：正解は赤色、誤答は青色

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← スコア・状態表示
├─────────────────┤
│   計算式表示    │ ← 奇数×偶数の内訳
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：1-1000
- デフォルト：16
- 説明：奇数スコア×偶数スコアがこの値以上で勝ち抜け

## 戦略的要素

### バランスの重要性

SquareX形式の最大の特徴は、奇数問目と偶数問目の両方でバランスよく正解する必要があることです：

1. **片方特化の限界**：一方のスコアが0だと最終スコアも0
2. **バランス戦略**：両方を均等に伸ばすことが基本
3. **タイミング戦略**：問題の難易度を見極めた解答判断

### 数学的特性

#### スコア増加パターンの分析

**16点勝ち抜けの場合の可能な組み合わせ**：

| 奇数スコア | 偶数スコア | 最終スコア |
| ---------- | ---------- | ---------- |
| 4          | 4          | 16         |
| 2          | 8          | 16         |
| 8          | 2          | 16         |
| 1          | 16         | 16         |
| 16         | 1          | 16         |

#### 効率性の分析

- **バランス型**：4×4 = 必要正解数8回
- **偏重型**：2×8 = 必要正解数10回
- **極端型**：1×16 = 必要正解数17回

バランスを保つほど効率的に勝ち抜けられる特性があります。

### プレイヤーの戦略

1. **バランス戦略**：奇数・偶数問目を均等に狙う
2. **得意分野戦略**：得意な分野の問題タイミングで積極解答
3. **効率重視戦略**：数学的に最短経路を計算して解答
4. **安全重視戦略**：確実な問題のみ解答してバランス維持

### 心理的要素

- **バランスプレッシャー**：片方が進みすぎることへの不安
- **機会損失の恐れ**：得意問題を見送ることの心理的負担
- **計算負荷**：常時スコア予測を行う必要性

## 他形式との比較

### nomx形式との違い

- **nomx**：単一次元での蓄積（正解数のみ）
- **squarex**：二次元での蓄積（奇数・偶数の両軸）

### nbyn形式との違い

- **nbyn**：正解数と誤答数の関係性
- **squarex**：奇数問目と偶数問目の関係性

### ny形式との違い

- **ny**：線形的なスコア蓄積
- **squarex**：乗法的なスコア蓄積

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 最終正解問題番号の早さ
4. 正解数（多い順）
5. 誤答数（少ない順）

### パフォーマンス考慮事項

- 奇数・偶数スコアの独立管理による効率的な計算
- 乗算処理による簡潔なスコア算出
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

### データ管理の特徴

- `odd_score`と`even_score`の独立したカウンター
- 問題番号による条件分岐の効率的な実装
- 失格条件がないシンプルな状態管理

## 使用例

### 設定例1：標準的なSquare16設定

- 勝ち抜けポイント：16
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦Square9設定

- 勝ち抜けポイント：9
- 用途：時間制限のある競技

### 設定例3：長期戦Square25設定

- 勝ち抜けポイント：25
- 用途：実力重視の長時間競技

### 設定例4：初心者向けSquare4設定

- 勝ち抜けポイント：4
- 用途：学習目的の競技

## 競技性の特徴

### 適用場面

1. **バランス感覚の評価**：二次元的な戦略立案能力
2. **数学的思考力**：効率的な組み合わせの計算能力
3. **機会判断力**：解答タイミングの最適化
4. **持続的集中力**：長期間にわたるバランス維持

### 出題者の配慮

- **分野バランス**：奇数・偶数問目での分野分布の調整
- **難易度調整**：両軸で均等に解答機会を提供
- **問題数設定**：十分な逆転機会の確保

## 実装上の課題とTODO

### 現在の制限事項

#### 1. バランス可視化の強化

**現状**：計算式表示のみ
**改善案**：奇数・偶数スコアの個別進捗表示

#### 2. 戦略ガイダンス

**現状**：数値情報のみ
**改善案**：最適解答タイミングのアドバイス

#### 3. 効率性の表示

**現状**：現在スコアのみ表示
**改善案**：勝ち抜けまでの最短経路の表示

### 将来的な改善提案

1. **バランス度の可視化**：奇数・偶数スコアの偏りの表示
2. **効率性計算機能**：最短勝ち抜け経路の自動計算
3. **統計機能**：各軸での正答率の分析表示
4. **学習支援機能**：バランス戦略の学習コンテンツ

この形式は、従来の一次元的なスコア蓄積とは異なる二次元的な戦略性を持つルールとして設計されており、プレイヤーに数学的思考とバランス感覚の両方を要求する独特な競技形式です。奇数問目と偶数問目という時間軸に基づく分類により、解答タイミングの戦略性を高めた革新的なルールです。
