# Swedish10形式仕様書

## 概要

Swedish10形式（スウェディッシュテン）は、競技クイズにおいて正答数に応じて誤答時のダメージポイントが変動する独特なルール形式です。10回の正答で勝ち抜けを目指しますが、誤答すると正答数に応じて異なる失点（1×、2×、3×、4×）が付与され、10×以上で失格となります。正答数が多いほど誤答時のリスクが高まるため、慎重さと積極性のバランスが重要となる戦略性の高い形式です。

## 基本ルール

### 勝利条件

- 10回正解することで勝ち抜け
- デフォルト設定：10回正解で勝ち抜け

### 失格条件

- 失点の合計が10×以上になると失格
- デフォルト設定：10×以上で失格

### 誤答時の失点システム

誤答時の失点は、誤答した時点での正答数によって決定されます：

- **正答数0回時の誤答**：1×
- **正答数1-2回時の誤答**：2×
- **正答数3-5回時の誤答**：3×
- **正答数6回以上時の誤答**：4×

### スコア表示

プレイヤーの状態は「正答数 - 失点数」の形式で表示されます（例："7 - 3"）

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "swedish10",
  win_point: number,    // 勝ち抜けに必要な正答数（デフォルト：10）
  lose_point: number,   // 失格となる失点数（デフォルト：10）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | swedish10形式での扱い  | 説明                                     |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子             |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名                       |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の正解数                     |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の誤答数                     |
| `base_correct_point` | `number` | 通常1                  | 使用されない           | 正解時の基本点数（swedish10では未使用）  |
| `base_wrong_point`   | `number` | 通常1                  | 使用されない           | 誤答時の基本点数（swedish10では未使用）  |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値      | 正解時           | 誤答時     | 説明                     |
| --------------- | --------- | ----------- | ---------------- | ---------- | ------------------------ |
| `score`         | `number`  | `0`         | 変更なし         | `失点数`   | 失点数の累計             |
| `correct`       | `number`  | `0`         | `+1`             | 変更なし   | 現在の正解数             |
| `wrong`         | `number`  | `0`         | 変更なし         | `+失点値`  | 現在の失点数             |
| `last_correct`  | `number`  | `-1`        | 問題番号         | 変更なし   | 最後に正解した問題番号   |
| `last_wrong`    | `number`  | `-1`        | 変更なし         | 問題番号   | 最後に誤答した問題番号   |
| `state`         | `string`  | `"playing"` | 勝利時`"win"`    | 失格時`"lose"` | プレイヤーの状態         |
| `reach_state`   | `string`  | `"playing"` | 変更なし         | 変更なし   | リーチ状態（使用されない） |
| `text`          | `string`  | 動的生成    | 動的更新         | 動的更新   | 表示テキスト             |
| `order`         | `number`  | `0`         | 勝利時に順位設定 | 変更なし   | 勝ち抜け順位             |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. 正解数が`win_point`（10回）に到達した場合、勝ち抜け状態に設定

```typescript
const newCorrect = playerState.correct + 1;
if (newCorrect >= game.win_point!) {
  // 勝ち抜け
  state = "win";
}
```

#### 誤答時の処理

1. 現在の正解数に基づいて失点値を決定
2. プレイヤーの失点数に失点値を加算
3. 失点数が`lose_point`（10×）に到達した場合、失格状態に設定

```typescript
let newWrong = playerState.wrong;
if (playerState.correct <= 0) {
  newWrong += 1;      // 0回正解時：1×
} else if (playerState.correct <= 2) {
  newWrong += 2;      // 1-2回正解時：2×
} else if (playerState.correct <= 5) {
  newWrong += 3;      // 3-5回正解時：3×
} else {
  newWrong += 4;      // 6回以上正解時：4×
}

if (newWrong >= game.lose_point!) {
  // 失格
  state = "lose";
}
```

#### 失点計算の詳細

| 正解数 | 誤答時失点 | 累積例 |
|--------|------------|--------|
| 0      | 1×         | 1×     |
| 1-2    | 2×         | 3×, 5× |
| 3-5    | 3×         | 6×, 9× |
| 6-9    | 4×         | 10×以上で失格 |

### UI表示仕様

#### スコアボード表示

2行構成でプレイヤー情報を表示：

- **1行目**：メイン状態表示
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時："LOSE"
  - 通常時：正解数（例："7pt"）

- **2行目**：正解・誤答操作ボタン（2個並び）
  - 左側：正解ボタン（赤色、コンパクト表示）
  - 右側：誤答ボタン（青色、コンパクト表示）
  - **失格後は両ボタンが無効化（disabled）される**

#### 色分け

- **メインボタン**：
  - 通常時：プレイヤーの状態に応じた色
  - 勝ち抜け時：勝利状態の色
  - 失格時：失格状態の色

- **操作ボタン**：
  - 正解は赤色、誤答は青色
  - 失格時はグレーアウト

#### レイアウト構成

```
┌─────────────────┐
│   メイン表示    │ ← 正解数・状態表示
├────────┬────────┤
│ 正解   │ 誤答   │ ← 操作ボタン
└────────┴────────┘
```

### 設定可能パラメータ

#### 勝ち抜け正解数（win_point）

- 範囲：1-1000
- デフォルト：10
- 説明：この回数正解すると勝ち抜け

#### 失格失点数（lose_point）

- 範囲：1-1000
- デフォルト：10
- 説明：この失点数以上で失格

## 戦略的要素

### 正解数によるリスク変動

Swedish10形式の最大の特徴は、正解数が増えるほど誤答時のリスクが高まることです：

1. **序盤（0-2回正解）**：比較的安全な誤答（1×-2×）
2. **中盤（3-5回正解）**：中程度のリスク（3×）
3. **終盤（6回以上正解）**：高リスク（4×）

### 数学的特性

#### 許容誤答回数の計算

各段階での理論的最大誤答回数：

| 正解数段階 | 失点値 | 最大誤答回数 | 累積失点 |
|------------|--------|--------------|----------|
| 0回正解時  | 1×     | 9回          | 9×       |
| 1-2回正解時| 2×     | 4回          | 8×       |
| 3-5回正解時| 3×     | 3回          | 9×       |
| 6回以上時  | 4×     | 2回          | 8×       |

#### 危険度分析

正解数が進むほど、1回の誤答が致命的になる特性：

- **0-2回正解**：比較的余裕がある段階
- **3-5回正解**：慎重さが要求される段階
- **6回以上正解**：極めて慎重な解答が必要

### プレイヤーの戦略

1. **段階的戦略**：正解数に応じた解答スタイルの変更
2. **リスク計算戦略**：現在の失点と許容誤答回数の把握
3. **確実性重視戦略**：6回以上正解後の超慎重解答
4. **早期決着戦略**：序盤での積極的な解答で先行を狙う

### 心理的要素

- **段階的プレッシャー**：正解数が増えるほど高まる緊張感
- **計算負荷**：現在のリスクレベルの常時把握
- **失点への恐怖**：高正解数時の誤答に対する不安

## 他形式との比較

### nomx形式との違い

- **nomx**：固定的な誤答による失格（例：3回誤答で失格）
- **swedish10**：変動的な失点システム（正解数に応じた失点変動）

### divide形式との違い

- **divide**：誤答回数による除算（累積的な減少効果）
- **swedish10**：正解数による失点変動（段階的なリスク増加）

### nupdown形式との違い

- **nupdown**：誤答によるスコアリセット（一律の厳しいペナルティ）
- **swedish10**：段階的な失点システム（リスクの段階的増加）

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在の正解数（多い順）
3. 最終正解問題番号の早さ
4. 現在の失点数（少ない順）

### パフォーマンス考慮事項

- 条件分岐による効率的な失点計算
- UI更新は状態変更時のみ実行され、不要な再描画を抑制
- 失格判定の明確な実装

### 状態管理の特徴

- `score`フィールドを失点数として再利用
- `wrong`フィールドに累積失点数を格納
- 失格後の操作制限の実装

## 使用例

### 設定例1：標準的なSwedish10設定

- 勝ち抜け正解数：10
- 失格失点数：10
- 用途：一般的な早押しクイズ大会

### 設定例2：短期戦設定

- 勝ち抜け正解数：7
- 失格失点数：10
- 用途：時間制限のある競技

### 設定例3：高難易度設定

- 勝ち抜け正解数：10
- 失格失点数：8
- 用途：よりシビアな競技

### 設定例4：寛容設定

- 勝ち抜け正解数：10
- 失格失点数：15
- 用途：学習目的の競技

## 競技性の特徴

### 適用場面

1. **段階的判断力の評価**：状況に応じた戦略変更能力
2. **リスク管理能力**：数学的なリスク計算能力
3. **プレッシャー耐性**：高正解数時の冷静な判断力
4. **持続的集中力**：長期間にわたる注意力維持

### 出題者の配慮

- **段階別問題調整**：正解数に応じた問題難易度の調整
- **リスク可視化**：現在のリスクレベルの分かりやすい表示
- **時間配分**：各段階での適切な思考時間の確保

## 実装上の課題とTODO

### 現在の制限事項

#### 1. リスクレベルの可視化

**現状**：基本的なスコア表示のみ
**改善案**：現在のリスクレベル（失点値）の明示的表示

#### 2. 戦略ガイダンス

**現状**：数値のみの表示
**改善案**：許容誤答回数や危険度の表示

#### 3. 段階遷移の表現

**現状**：数値変化のみ
**改善案**：リスクレベル変更時の視覚的フィードバック

### 将来的な改善提案

1. **リスク計算機能**：現在の許容誤答回数の自動表示
2. **段階表示機能**：現在のリスクレベルの色分け表示
3. **統計機能**：各段階での正答率・誤答率の分析
4. **学習支援機能**：最適戦略のアドバイス表示

この形式は、競技の進行とともにリスクが段階的に変化する独特な設計により、プレイヤーに状況判断力と数学的思考力の両方を要求する高度な戦略性を持つルールです。正解数の増加とともに高まるプレッシャーにより、技術力だけでなく精神力も重要となる競技形式として設計されています。