# Variables形式仕様書

## 概要

Variables形式は、競技クイズにおいて各プレイヤーが個別の変動値を設定し、その値に応じて正答・誤答時のスコア変動が決まるルール形式です。プレイヤーが戦略的にリスクとリターンのバランスを選択できる特徴があります。

## 基本ルール

### 勝利条件

- 設定されたスコア（勝ち抜けポイント）に到達することで勝ち抜け
- デフォルト設定：30ポイントで勝ち抜け

### スコア変動システム

- 各プレイヤーは個別に「変動値」を設定可能
- **正解時**：設定した変動値分のポイントを獲得
- **誤答時**：設定した減点値分のポイントを失う
- 失格条件は存在せず、全プレイヤーが最後まで参加可能

### 変動値の設定

- プレイヤーごとに自由に数値を設定可能
- 制限値なし（負の値も設定可能）
- ゲーム開始前のconfig画面で設定

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "variables",
  win_point: number,    // 勝ち抜けに必要なスコア（デフォルト：30）
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | Variables形式での扱い  | 説明                         |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子 |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名           |
| `initial_correct`    | `number` | 通常0（調整可能）      | スコア初期値として使用 | ゲーム開始時の正解数         |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 計算に影響なし         | ゲーム開始時の誤答数         |
| `base_correct_point` | `number` | 通常10（調整可能）     | 正解時の獲得ポイント   | 正解時の基本点数             |
| `base_wrong_point`   | `number` | 通常-10（調整可能）    | 誤答時の減点ポイント   | 誤答時の基本点数             |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値      | 正解時                       | 誤答時                     | 説明                       |
| --------------- | --------- | ----------- | ---------------------------- | -------------------------- | -------------------------- |
| `score`         | `number`  | `0`         | `score + base_correct_point` | `score + base_wrong_point` | 現在のスコア               |
| `correct`       | `number`  | `0`         | `+1`                         | 変更なし                   | 現在の正解数               |
| `wrong`         | `number`  | `0`         | 変更なし                     | `+1`                       | 現在の誤答数               |
| `last_correct`  | `number`  | `-10`       | 問題番号                     | 変更なし                   | 最後に正解した問題番号     |
| `last_wrong`    | `number`  | `-10`       | 変更なし                     | 問題番号                   | 最後に誤答した問題番号     |
| `is_incapacity` | `boolean` | `false`     | 変更なし                     | 変更なし                   | 解答権剥奪フラグ（未使用） |
| `state`         | `string`  | `"playing"` | 勝利時`"win"`                | 変更なし                   | プレイヤーの状態           |
| `reach_state`   | `string`  | `"normal"`  | リーチ時`"win"`              | 変更なし                   | リーチ状態                 |
| `text`          | `string`  | 動的生成    | 動的更新                     | 動的更新                   | 表示テキスト               |
| `order`         | `number`  | `0`         | 勝利時に順位設定             | 変更なし                   | 勝ち抜け順位               |

### 計算ロジック

#### 正答時の処理

1. プレイヤーの正解数を1増加
2. スコアに`base_correct_point`を加算
3. スコアが`win_point`に到達した場合、勝ち抜け状態に設定
4. スコアが次回正解で`win_point`に到達する場合、勝ち抜けリーチ状態に設定

```typescript
const correct_point =
  game.players.find((gamePlayer) => gamePlayer.id === playerState.player_id)
    ?.base_correct_point || 0;
const newScore = playerState.score + correct_point;
```

#### 誤答時の処理

1. プレイヤーの誤答数を1増加
2. スコアに`base_wrong_point`を加算（通常は負の値なので実質減算）
3. `last_wrong`に現在の問題番号を記録

```typescript
const wrong_point =
  game.players.find((gamePlayer) => gamePlayer.id === playerState.player_id)
    ?.base_wrong_point || 0;
return {
  ...playerState,
  wrong: playerState.wrong + 1,
  score: playerState.score + wrong_point,
  last_wrong: qn,
};
```

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 通常時：現在のスコア（例："15pt"）

- **サブ表示部分**
  - 正解数と誤答数の表示
  - 変動値表示（正解時ポイント/誤答時ポイント）

- **色分け**
  - 通常時：緑色
  - 勝ち抜け時：金色

#### レイアウト構成

3行構成でプレイヤー情報を表示：

- 上段：メインスコア表示
- 中段：正解・誤答数表示
- 下段：変動値表示（+N/-M形式）

### 設定可能パラメータ

#### 勝ち抜けポイント（win_point）

- 範囲：3-1000
- デフォルト：30
- 説明：このスコアに到達すると勝ち抜け

#### プレイヤー個別設定

- **正解時ポイント**（base_correct_point）
  - 範囲：制限なし
  - デフォルト：10
  - 説明：正解時に獲得するポイント

- **誤答時ポイント**（base_wrong_point）
  - 範囲：制限なし
  - デフォルト：-10
  - 説明：誤答時に失うポイント（負の値で設定）

## 戦略的要素

### プレイヤーの戦略

1. **安全重視戦略**：低い変動値を設定してリスクを抑制
2. **ハイリスク・ハイリターン戦略**：高い変動値を設定して一気に上位を狙う
3. **バランス戦略**：中程度の変動値で安定した進行を目指す
4. **逆転戦略**：劣勢時に高い変動値に変更して逆転を狙う

### 設定例による戦略比較

| 戦略タイプ | 正解時 | 誤答時 | リスク | 特徴                   |
| ---------- | ------ | ------ | ------ | ---------------------- |
| 安全重視   | +5     | -3     | 低     | 確実な進行が可能       |
| 標準       | +10    | -10    | 中     | バランスの取れた設定   |
| 攻撃的     | +15    | -20    | 高     | 早期勝利と大失速の両面 |

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のスコア（高い順）
3. 正解数（多い順）
4. 誤答数（少ない順）
5. 最終正解問題番号（早い順）

### パフォーマンス考慮事項

- 各プレイヤーの変動値は個別に管理され、リアルタイムでスコア計算を実行
- UI更新は状態変更時のみ実行され、不要な再描画を抑制

## 使用例

### 設定例1：入門者向け設定

- 勝ち抜けポイント：20
- 正解時ポイント：+8
- 誤答時ポイント：-5
- 用途：初心者向けの競技

### 設定例2：標準的な設定

- 勝ち抜けポイント：30
- 正解時ポイント：+10
- 誤答時ポイント：-10
- 用途：一般的な競技クイズ大会

### 設定例3：上級者向け設定

- 勝ち抜けポイント：50
- 正解時ポイント：+15
- 誤答時ポイント：-20
- 用途：ハイレベルな競技

## 他形式との比較

### Normal形式との違い

- **Normal**：固定的なスコア変動
- **Variables**：プレイヤーごとに個別の変動値設定が可能

### Swedish形式との違い

- **Swedish**：正解数ベースの勝利条件
- **Variables**：スコアベースの勝利条件

### AQL形式との違い

- **AQL**：複雑な得点計算式
- **Variables**：シンプルな加減算によるスコア管理

## 実装上の特徴

### カスタマイズ性

Variables形式の最大の特徴は、プレイヤーごとに異なる戦略を採用できることです。同一ゲーム内で：

- 保守的なプレイヤーは低リスク設定
- 積極的なプレイヤーは高リスク設定
- 状況に応じた戦略変更（ゲーム開始前のみ）

### 教育的価値

この形式は以下の教育的効果があります：

1. **リスク管理の理解**：リスクとリターンの関係性を学習
2. **戦略的思考の育成**：状況に応じた最適解の選択
3. **確率論の実践**：期待値計算の重要性の理解

このVariables形式は、従来の固定的なルールを超えて、プレイヤーの戦略性と個性を最大限に活かせる革新的なゲーム形式として設計されています。

## ユニットテスト観点

ユニットテストでは`getInitialPlayersStateForOnline`がvariables形式に対してスコア0、正解数0で初期化しつつ、プレイヤーごとの`initialScore`が誤答数の初期値としてのみ適用されることを確認する。異なる`initialScore`を与えた複数プレイヤーを初期化し、正解数は常に0で開始する一方で誤答数が値を引き継ぐこと、全員が`reach_state: "playing"`のまま並び替え処理に渡されること、そしてスコアが同一の場合に正解数・誤答数による順位付けが適切に行われることを検証する。
