# Z形式仕様書

## 概要

Z形式は、競技クイズにおいて5つのステージを順次攻略することで勝ち抜けを目指すルール形式です。各ステージでは要求される正解数と許容誤答数が段階的に厳しくなり、誰かがステージをクリアした時点で全員の成績がリセットされる独特のシステムを採用しています。

## 基本ルール

### 勝利条件

- ステージ5に到達することで勝ち抜け
- ステージ4で4回正解することでステージ5へ進行

### ステージ構成

各ステージには固有のクリア条件と失格条件が設定されています：

| ステージ | クリア条件 | 失格条件 | 特記事項 |
|----------|------------|----------|----------|
| ステージ1 | 1回正解 | なし | 1回誤答で1問解答権剥奪 |
| ステージ2 | 2回正解 | 1回誤答 | - |
| ステージ3 | 3回正解 | 2回誤答 | - |
| ステージ4 | 4回正解 | 3回誤答 | クリアで勝ち抜け |

### リセット機能

いずれかのプレイヤーがステージをクリアした時点で、以下の処理が実行されます：

1. **クリアしたプレイヤー**：次のステージに進行
2. **その他のプレイヤー**：現在ステージの正解数・誤答数・失格状態がリセットされ、同ステージで再スタート

この仕組みにより、常に誰かが新しいステージに挑戦している状況が維持されます。

## 実装仕様

### データ構造

#### ゲーム設定

```typescript
{
  rule: "z",
  // Z形式では設定可能なパラメータなし
  // ステージ構成は固定
}
```

#### プレイヤー状態

Score Watcherでは、プレイヤーデータは3つの段階で管理されています：

1. **マスターデータ**（`players`テーブル）：プレイヤーの基本情報
2. **ゲーム内データ**（`GameDBPlayerProps`）：ゲーム固有の設定
3. **計算済みスコア**（`ComputedScoreProps`）：リアルタイムで計算される状態

##### マスターデータ（PlayerDBProps）

| プロパティ名 | 型         | 初期値       | 説明                           |
| ------------ | ---------- | ------------ | ------------------------------ |
| `id`         | `string`   | 自動生成UUID | プレイヤーの一意識別子         |
| `name`       | `string`   | ユーザー入力 | プレイヤー名                   |
| `text`       | `string`   | ユーザー入力 | プレイヤーの説明文・備考       |
| `belong`     | `string`   | ユーザー入力 | 所属団体・学校など             |
| `tags`       | `string[]` | `[]`         | プレイヤーに付与するタグの配列 |

##### ゲーム内データ（GameDBPlayerProps）

| プロパティ名         | 型       | 初期値                 | Z形式での扱い          | 説明                               |
| -------------------- | -------- | ---------------------- | ---------------------- | ---------------------------------- |
| `id`                 | `string` | マスターデータより継承 | プレイヤー識別に使用   | ゲーム内でのプレイヤー識別子       |
| `name`               | `string` | マスターデータより継承 | 表示に使用             | ゲーム内での表示名                 |
| `initial_correct`    | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の正解数               |
| `initial_wrong`      | `number` | 通常0（調整可能）      | 使用されない           | ゲーム開始時の誤答数               |
| `base_correct_point` | `number` | 通常10                 | 使用されない           | 正解時の基本点数（Z形式では未使用） |
| `base_wrong_point`   | `number` | 通常-10                | 使用されない           | 誤答時の基本点数（Z形式では未使用） |

##### 計算済みスコア（ComputedScoreProps）

| プロパティ名    | 型        | 初期値      | 正解時                   | 誤答時                     | リセット時      | 説明                         |
| --------------- | --------- | ----------- | ------------------------ | -------------------------- | --------------- | ---------------------------- |
| `score`         | `number`  | `0`         | 変更なし                 | 変更なし                   | 変更なし        | Z形式では使用されない        |
| `correct`       | `number`  | `0`         | `+1`                     | 変更なし                   | `0`             | 現在ステージでの正解数       |
| `wrong`         | `number`  | `0`         | 変更なし                 | `+1`                       | `0`             | 現在ステージでの誤答数       |
| `stage`         | `number`  | `1`         | クリア時`+1`             | 変更なし                   | 変更なし        | 現在のステージ番号           |
| `last_correct`  | `number`  | `-10`       | 問題番号                 | 変更なし                   | 変更なし        | 最後に正解した問題番号       |
| `last_wrong`    | `number`  | `-10`       | 変更なし                 | 問題番号                   | `-10`           | 最後に誤答した問題番号       |
| `is_incapacity` | `boolean` | `false`     | 変更なし                 | 失格時または解答権剥奪時`true` | `false`         | 解答権剥奪・失格状態フラグ   |
| `state`         | `string`  | `"playing"` | ステージ5到達時`"win"`   | 失格時`"lose"`             | `"playing"`     | プレイヤーの状態             |
| `reach_state`   | `string`  | `"normal"`  | 動的判定                 | 動的判定                   | `"normal"`      | リーチ状態                   |
| `text`          | `string`  | `"Stage1"`  | 動的更新                 | 動的更新                   | 動的更新        | 表示テキスト                 |
| `order`         | `number`  | `0`         | 勝利時に順位設定         | 変更なし                   | 変更なし        | 勝ち抜け順位                 |

### 計算ロジック

#### 正答時の処理

```typescript
// 1. 正解数を1増加
const newCorrect = playerState.correct + 1;

// 2. ステージクリア判定
if (stage === newCorrect) {
  // ステージクリア：次ステージへ進行し、成績をリセット
  return {
    ...playerState,
    correct: 0,
    wrong: 0,
    stage: playerState.stage + 1,
    is_incapacity: false,
    last_correct: qn,
    state: playerState.stage + 1 === 5 ? "win" : "playing",
  };
}

// 3. ステージ5到達で勝ち抜け
if (playerState.stage + 1 === 5) {
  return { ...playerState, state: "win" };
}
```

#### 誤答時の処理

```typescript
const newWrong = playerState.wrong + 1;

// ステージ1の特別処理：失格ではなく解答権剥奪のみ
if (stage === 1 && newWrong === 1) {
  return {
    ...playerState,
    wrong: 0,  // 誤答数はリセット
    last_wrong: qn,
    is_incapacity: true,
  };
}

// ステージ2以降の失格判定
if (stage === newWrong + 1) {
  return {
    ...playerState,
    wrong: newWrong,
    is_incapacity: true,
    last_wrong: qn,
    state: "lose",
  };
}
```

#### リセット処理

```typescript
// 他のプレイヤーがステージクリア時
if (isResetTiming && playerState.state !== "win") {
  return {
    ...playerState,
    correct: 0,
    wrong: 0,
    is_incapacity: false,
    state: "playing",
    last_wrong: -10,  // 解答権剥奪状態もリセット
  };
}
```

#### 解答権剥奪の回復判定

ステージ1での誤答による解答権剥奪は、1問経過後に自動回復します：

```typescript
// 解答権回復条件
const isRecovered = gameLogList.length > playerState.last_wrong;
```

### UI表示仕様

#### スコアボード表示

- **メイン表示部分**
  - 勝ち抜け時：順位（"1st", "2nd", "3rd", "4th", ...）
  - 失格時または解答権剥奪時："休"
  - 通常時：現在のステージ（"Stage1", "Stage2", "Stage3", "Stage4"）

- **操作ボタン**
  - 正解ボタン：失格時・解答権剥奪時はグレーアウトして無効化
  - 誤答ボタン：失格時・解答権剥奪時はグレーアウトして無効化

- **色分け**
  - メインボタン：失格・解答権剥奪状態では青色、通常時は緑色
  - 勝ち抜け時：金色

#### レイアウト構成

2行構成でプレイヤー情報を表示：

- 上段：メイン状態表示（ステージ番号または順位・休み状態）
- 下段：正解・誤答操作ボタン

### 設定可能パラメータ

Z形式では設定可能なパラメータはありません。ステージ構成とルールはすべて固定されています。

## 他形式との比較

### nomx形式との違い

- **nomx**：一定回数の誤答で即失格、個人戦形式
- **Z**：ステージごとに許容誤答数が変化、リセット機能による集団戦形式

### nomr形式との違い

- **nomr**：誤答で一時的な解答権剥奪、失格なし
- **Z**：ステージによって誤答の影響が変化、リセット機能あり

### swedish形式との違い

- **swedish**：ポイント減算による失格システム
- **Z**：ステージクリア型の進行システム

## 戦略的要素

### プレイヤーの戦略

1. **早期クリア戦略**：自分がステージクリアすることで他プレイヤーをリセットし、先行を維持
2. **慎重戦略**：他プレイヤーのクリアを待ち、リセットのタイミングで確実に進行
3. **追い上げ戦略**：遅れているステージで集中的に解答し、一気に追い付く

### 出題者の配慮

- ステージクリア後のリセットタイミングでの問題調整
- 各ステージの特性を活かした問題難易度の設定
- 失格者が多い場合の進行方法の検討

## 技術的な特徴

### ソート順序

プレイヤーの順位は以下の優先順位で決定されます：

1. 勝ち抜け状態（勝ち抜け順）
2. 現在のステージ番号
3. 現在ステージでの正解数
4. 最終正解問題番号の早さ
5. 誤答数（少ない方が上位）

### パフォーマンス考慮事項

- ステージクリア時の全体リセット処理により、状態管理が簡潔
- 各ステージでの個別状態管理により、効率的な進行判定を実現

## 使用例

### 設定例：固定設定

Z形式では設定項目がないため、すべてのゲームで同じルールが適用されます：

- ステージ1：1正解でクリア、1誤答で解答権剥奪
- ステージ2：2正解でクリア、1誤答で失格
- ステージ3：3正解でクリア、2誤答で失格
- ステージ4：4正解でクリア、3誤答で失格

このルールは、段階的に難易度が上がるとともに、リセット機能により常に緊張感のある競技を提供します。

## 実装上の課題とTODO

### バリデーション強化が必要な項目

#### 1. 失格・解答権剥奪プレイヤーの操作防止

**現状**：クライアント側でボタンを無効化しているが、サーバーサイドでのバリデーションが不十分
**対策**：失格状態（`state: "lose"`）または解答権剥奪状態（`is_incapacity: true`）のプレイヤーからの解答操作を受け付けないよう、バックエンドレベルでのバリデーションを追加する必要がある

#### 2. 勝ち抜け後プレイヤーの操作制限

**現状**：勝ち抜け後も操作が可能な状態
**対策**：勝ち抜け状態（`state: "win"`）のプレイヤーからの操作を無効化するバリデーションを実装する必要がある

#### 3. リセットタイミングの整合性

**現状**：複数プレイヤーが同時にステージクリアした場合の処理が不明確
**対策**：同時クリア時の優先順位や処理順序を明確に定義する必要がある

### 将来的な改善提案

1. **ステージ表示の強化**：現在のステージ進行状況をより視覚的に表示
2. **リセット通知機能**：他プレイヤーのクリアによるリセットを明確に通知
3. **ステージ履歴の記録**：各プレイヤーのステージクリア履歴を追跡
4. **カスタマイズ機能の検討**：将来的にステージ構成を変更可能にするかの検討